In [1]:
# %load std_libraries.py
#to keep my scripts consistent, I am adding this to the top of all scripts with %load std_libraries.py
import sys
#Indicate operating environment and import core modules
location_input = input("what computer are you on? a = Bens, b = gpucluster, c = other   ")
location_dict = {'a': "C:\\Users\\heine\\github\\expression_broad_data", 'b': "/home/heineike/github/expression_broad_data",'c':'you need to add your location to the location_dict'}
base_dir = location_dict[location_input]
print("base directory is " + base_dir)

if sys.path[-1] != base_dir:
    sys.path.append(base_dir)
    print("Added " + base_dir + " to path: " )
    print(sys.path)

import os

print("I am about to import a library")
from core import expression_plots 
from core import io_library 
%load_ext autoreload
%autoreload 2
# %matplotlib 
# plt.ioff()
#%matplotlib inline
%matplotlib notebook
data_processing_dir = base_dir + os.sep + os.path.normpath("expression_data") + os.sep

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
#import seaborn as sns; sns.set(style="ticks", color_codes=True)
#from sklearn import linear_model
#import pickle
#import subprocess
#import networkx as nx
# import scipy.stats as stats
import scipy.cluster.hierarchy as sch

# from Bio import SeqIO
# from Bio import SeqFeature as sf
# from Bio.Alphabet import generic_dna
# from Bio.Seq import Seq

# import re

from collections import Counter
# import scipy.stats as stats
# from itertools import chain
#this only works if you are online
online_input = input("are you online? Yes/No")
if online_input == "Yes": 
    import plotly.plotly as py
    import plotly.graph_objs as pygo
    import plotly.tools as pytools
    py.sign_in('heineike02_student','9dMTMZgJMgUP0YX0P5mQ')
    #py.sign_in('heineike02', 'APPjKrtARaN2ZgUYIkqr')
    
# for phylogenetic trees: 
from ete3 import Tree


what computer are you on? a = Bens, b = gpucluster, c = other   a
base directory is C:\Users\heine\github\expression_broad_data
Added C:\Users\heine\github\expression_broad_data to path: 
['', 'C:\\Users\\heine\\Anaconda3\\python36.zip', 'C:\\Users\\heine\\Anaconda3\\DLLs', 'C:\\Users\\heine\\Anaconda3\\lib', 'C:\\Users\\heine\\Anaconda3', 'C:\\Users\\heine\\Anaconda3\\lib\\site-packages', 'C:\\Users\\heine\\Anaconda3\\lib\\site-packages\\Babel-2.5.0-py3.6.egg', 'C:\\Users\\heine\\Anaconda3\\lib\\site-packages\\win32', 'C:\\Users\\heine\\Anaconda3\\lib\\site-packages\\win32\\lib', 'C:\\Users\\heine\\Anaconda3\\lib\\site-packages\\Pythonwin', 'C:\\Users\\heine\\Anaconda3\\lib\\site-packages\\IPython\\extensions', 'C:\\Users\\heine\\.ipython', 'C:\\Users\\heine\\github\\expression_broad_data']
I am about to import a library
I am importing io_library
what computer are you on? a = Bens, b = gpucluster, c = other   a
base directory is C:\Users\heine\github\expression_broad_data
data process

## Datasets to load

In [3]:
#Load KL SC PKA data
fname_cross_spec_comparison = data_processing_dir + os.path.normpath("20170817_klscpka/cross_species_compare_ASminus_ASplus_YPD.csv")
kl_sc_PKA_data = pd.read_csv(fname_cross_spec_comparison, index_col=0)


#Total number of KL genes that have a paralog in SCER
# kl_sc_PKA_data_paralogs = kl_sc_PKA_data[kl_sc_PKA_data["N SCer Orthologs"]==2]

# print('Total number of KL genes that map to a set of paralogs in S.Cer: ' + str(int(len(kl_sc_PKA_data_paralogs)/2)))

# #Total of these that are WGD paralogs
# ohnolog_set = set(ohnologs['Gene 1']) | set(ohnologs['Gene 2'])

# non_ohnologs = set(kl_sc_PKA_data_paralogs['sc_genename']) - ohnolog_set
# ohnologs_klac_orth = set(kl_sc_PKA_data_paralogs['sc_genename']) & ohnolog_set

# print('KL genes that map to a set of Ohnologs: ' + str(int(len(ohnologs_klac_orth)/2)) + '\n' +
#      'KL genes that map to a non Ohnolog paralog: ' + str(int(len(non_ohnologs))))


In [23]:
#Load Ohnolog data
ohnologs_expression_sorted = pd.read_csv(data_processing_dir + os.path.normpath('20170817_klscpka/20170817_ohnologs_sorted.csv'),index_col = 0)
ohnologs_expression_sorted

Number   Ancestor AA %id  Length Ratio SC_common_name_YGOB_low  \
0         1   Anc_1.15    25%          0.73                    HLR1   
1         2   Anc_1.17    60%          0.99                    APA1   
2         3   Anc_1.22    48%          0.84               YDR524C-B   
3         4   Anc_1.25    42%          0.92                    SPS2   
4         5   Anc_1.31    39%          0.99                    EUG1   
5         6   Anc_1.33    73%          1.00                    EMI2   
6         7   Anc_1.36    25%          0.97                    SRO9   
7         8   Anc_1.37    37%          0.85                    GFD2   
8         9   Anc_1.38    62%          0.77                    GRX1   
9        10   Anc_1.50    67%          0.95                    GNP1   
10       11   Anc_1.52    47%          0.91                    GIN4   
11       12   Anc_1.55    48%          0.75                    PSP1   
12       13   Anc_1.58    91%          0.99                    SAM1   
13       14   Anc_1.61    31%          0.94                    TOS4   
14       15   Anc_1.62    93%          1.00                  RPL37A   
15       16   Anc_1.66    32%          0.97                    SKG3   
16       17   Anc_1.83    65%          0.97                    CAN1   
17       18  Anc_1.100    37%          1.00                    YSP1   
18       19  Anc_1.136    69%          0.95                   PHO90   
19       20  Anc_1.140    55%          0.89                    FEN1   
20       21  Anc_1.146    97%          0.99                  RPS14A   
21       22  Anc_1.157    23%          0.82                 YJR030C   
22       23  Anc_1.161    98%          1.00                  RPL17B   
23       24  Anc_1.180    39%          0.85                    HAL5   
24       25  Anc_1.182    71%          1.00                    TPK3   
25       26  Anc_1.187    55%          0.84                    PIR1   
26       27  Anc_1.188    74%          0.79                  HSP150   
27       28  Anc_1.209    60%          0.99                    KTR2   
28       29  Anc_1.211   100%          1.00                    TIF2   
29       30  Anc_1.212    44%          0.62                    GLG2   
..      ...        ...    ...           ...                     ...   
489     514  Anc_8.566    28%          0.67                    SMP1   
490     515  Anc_8.567   100%          1.00                   RPS6B   
491     516  Anc_8.574    58%          0.99                    EEB1   
492     517  Anc_8.587    29%          0.87                    SMY2   
493     518  Anc_8.590    75%          1.00                    SSE1   
494     519  Anc_8.601    18%          0.95                   PEX25   
495     520  Anc_8.614    69%          0.98                    DED1   
496     522  Anc_8.644    62%          0.99                    ODC2   
497     523  Anc_8.647    70%          0.95                    ISU2   
498     524  Anc_8.648    44%          0.98                    GIP3   
499     525  Anc_8.651    37%          0.99                    WTM2   
500     526  Anc_8.652    57%          1.00                    MKK1   
501     527  Anc_8.654    41%          0.92                    KIN4   
502     528  Anc_8.655    99%          1.00                  RPL33A   
503     529  Anc_8.658    69%          1.00                    KES1   
504     530  Anc_8.687    28%          0.81                    SRL1   
505     531  Anc_8.702    44%          0.97                    TRE1   
506     532  Anc_8.760    97%          1.00                  RPS10B   
507     533  Anc_8.764    37%          0.99                   UAF30   
508     534  Anc_8.772    56%          0.97                    BUD7   
509     536  Anc_8.792    55%          0.99                    ZRC1   
510     537  Anc_8.796    61%          0.99                    FAA4   
511     539  Anc_8.813    53%          0.96                    TPS3   
512     540  Anc_8.816    23%          0.58                    CUE4   
513     541  Anc_8.827    73%

In [30]:
# Load goi data from goi data file created in paralog analysis script. 
goi_data = pd.read_csv(data_processing_dir + os.sep + "20170817_klscpka" + os.sep + "goi_data.csv")

#subset ohnologs_sorted to include only goi_data

ohnologs_goi = ohnologs_expression_sorted[ohnologs_expression_sorted['sc_genename_low'].isin(goi_data['sc_genename']) |
                                          ohnologs_expression_sorted['sc_genename_high'].isin(goi_data['sc_genename'])]

ortholog_dataset_goi = ohnologs_goi.loc[:,['sc_genename_low','sc_genename_high']]
ortholog_dataset_goi.columns = ['low', 'high']
ortholog_dataset_goi

# ## Should redo this to incorporate latest goi info
# ohnologs_goi = pd.read_csv(data_processing_dir + os.path.normpath('20170817_klscpka/20170817_ohnologs_goi.csv'),index_col = 0)
# ohnologs_goi

low       high
5            YDR516C    YCL040W
24           YKL166C    YJL164C
29           YJL137C    YKR058W
50           YOL024W    YFR017C
54           YLR258W    YFR015C
60           YCR005C    YNR001C
66           YJR048W    YEL039C
67           YJR049C    YEL041W
92           YOR108W    YNL104C
93           YOR101W    YNL098C
100          YNL069C    YIL133C
106          YNL053W    YIL113W
107          YNL052W    YIL111W
114          YMR145C    YDL085W
122          YKL141W    YMR118C
126          YKL127W    YMR105C
131          YMR081C    YKL093W
138          YKL035W    YHL012W
139          YKL038W    YBR033W
143          YMR037C    YKL062W
167          YOL059W    YDL022W
168          YOL056W    YDL021W
172          YDR001C    YBR001C
189          YPR074C    YBR117C
191          YBR121C    YPR081C
199          YPR145W    YGR124W
206          YER027C    YGL208W
213          YGL253W    YFR053C
218  Scer_YGOB_SDC25    YLR310C
220          YGL031C    YGR148C
..               ...        ...
259          YHR174W    YGR254W
260          YHR163W    YGR248W
261          YHR162W    YGR243W
276          YKL148C    YJL045W
280          YDR300C    YHR033W
285          YDR341C    YHR091C
286          YDR345C    YHR096C
299          YML022W    YDR441C
303          YML028W    YDR453C
331          YLR273C    YOR178C
339          YBR218C    YGL062W
341          YBR238C    YGL107C
360          YMR182C    YPL230W
375        YCR073W-A    YNR034W
379          YOR375C    YAL062W
383          YAL038W    YOR347C
390          YOL045W    YAL017W
398          YIL045W    YER054C
402          YIL053W    YER062C
404          YIL057C    YER067W
405          YER070W    YIL066C
410          YDL181W  YDL130W-A
419          YBL054W    YER088C
422          YBL061C    YER096W
437          YBL039C    YJR103W
444          YLR061W  YFL034C-A
445          YFL018C    YPL017C
470          YDR178W    YLR164W
511          YMR261C    YML100W
513          YML106W    YMR271C

[65 rows x 2 columns]

In [25]:
#Phylomedb convert and trees

sc_phylomedb_convert = pd.read_table(data_processing_dir + os.path.normpath("phylogenetic_trees/sc_id_conversion.txt"), header=None)
sc_phylomedb_convert.columns = ["phylomedb_name","sc_genename"]

trees = pd.read_table(data_processing_dir + os.path.normpath("phylogenetic_trees/best_trees.txt"), header = None)
trees.columns = ["phylomedb_name","alignment_method","score","tree"]
trees = pd.merge(trees, sc_phylomedb_convert, on = "phylomedb_name", how = 'left')
trees.set_index("phylomedb_name", inplace=True)


In [60]:
trees_reduced = pd.read_table(data_processing_dir + os.path.normpath("phylogenetic_trees/reduced_phylomes/Trees_STK.txt"), header = None)


0                                                  1
0     P43545_YEAST  (O14027_SCHPO:0.108174,(((Wican1_30969_HANAN:0...
1     P53730_YEAST  (Q9USD4_SCHPO:1.16015,(YALI0C02475g_YARLI:1.05...
2     P53285_YEAST  (YALI0E06083g_YARLI:0.485661,(((Dekbr1_21723_D...
3     P26263_YEAST  (P36620_SCHPO:0.226098,((YALI0C00253g_YARLI:0....
4     P53879_YEAST  (Q01112_SCHPO:0.0533211,(YALI0B15752g_YARLI:0....
5     P40459_YEAST  (Q09673_SCHPO:1.32444,(((Dekbr1_48826_DEKBR:1....
6     P25635_YEAST  (Q9C1X1_SCHPO:0.282875,(((((Wican1_51974_HANAN...
7     P23641_YEAST  (Q9P7V8_SCHPO:0.255146,(Dekbr1_19034_DEKBR:0.5...
8     P32906_YEAST  (Q9P7C3_SCHPO:0.629631,(((Dekbr1_50624_DEKBR:0...
9     P38065_YEAST  (Q9UU81_SCHPO:0.399787,(((Wican1_84870_HANAN:0...
10    P32477_YEAST  (Q09768_SCHPO:0.392537,(((Wican1_78579_HANAN:0...
11    P0CF18_YEAST  (Q09740_SCHPO:0.34295,((((Wican1_33342_HANAN:0...
12    Q03289_YEAST  (Q9C0W9_SCHPO:0.90766,((YALI0E01804g_YARLI:1.0...
13    Q04418_YEAST  (Dekbr1_3985_DEKBR:0.871358,((((TDEL0C06830_TO...
14    Q12325_YEAST  (Q9URY8_SCHPO:0.265986,(O74377_SCHPO:0.379845,...
15    Q02260_YEAST  (O42661_SCHPO:0.342878,(YALI0A04961g_YARLI:0.2...
16    P33204_YEAST  (Q92352_SCHPO:0.182139,(YALI0D02805g_YARLI:0.1...
17    P38966_YEAST  (YALI0D13156g_YARLI:0.919844,(Wican1_65019_HAN...
18    P53077_YEAST  (Wican1_62630_HANAN:0.716484,(TDEL0D06100_TORD...
19    P12689_YEAST  (O94623_SCHPO:1.14132,(((Q59Q77_CANAL:0.851582...
20    Q2V2P4_YEAST  (Q10167_SCHPO:0.542482,(YALI0F21241g_YARLI:0.2...
21    Q12069_YEAST  (Q09709_SCHPO:0.342682,((((Q59YK3_CANAL:0.6070...
22    O75012_YEAST  (Wican1_33845_HANAN:0.555205,(O75012_YEAST:0.1...
23    Q04311_YEAST  (O74977_SCHPO:0.855115,(YALI0E26081g_YARLI:1.7...
24    P27514_YEAST  (O59712_SCHPO:0.648068,(YALI0E31064g_YARLI:0.8...
25    P40456_YEAST  (Dekbr1_17432_DEKBR:0.484426,(((TDEL0B02140_TO...
26    P30665_YEAST  (P30666_SCHPO:0.476195,(YALI0F21945g_YARLI:1.0...
27    P00572_YEAST  (P36590_SCHPO:0.484249,(((((Q6CW62_KLULA:0.747...
28    P25657_YEAST  (O42930_SCHPO:0.747255,(Wican1_20174_HANAN:0.8...
29    Q03079_YEAST  (YALI0C10736g_YARLI:0.651605,(((TDEL0B01480_TO...
...            ...                                                ...
3901  Q86ZR7_YEAST  (O14165_SCHPO:0.365541,((((Dekbr1_50083_DEKBR:...
3902  P30952_YEAST  (YALI0D19140g_YARLI:0.377795,(Dekbr1_47213_DEK...
3903  Q04002_YEAST  (Q09725_SCHPO:1.28544,(YALI0A14707g_YARLI:1.70...
3904  P39968_YEAST  (O43028_SCHPO:0.256665,(YALI0E13992g_YARLI:0.2...
3905  P38731_YEAST  (Q92341_SCHPO:0.299128,((Wican1_24870_HANAN:0....
3906  Q08774_YEAST  (Dekbr1_32630_DEKBR:0.69369,(Q5ACQ4_CANAL:1.46...
3907  P04397_YEAST  (Q9HDU3_SCHPO:0.222649,(Dekbr1_24416_DEKBR:0.8...
3908  Q12202_YEAST  (YALI0D20438g_YARLI:0.948861,(Wican1_77026_HAN...
3909  P36039_YEAST  (Q9P787_SCHPO:0.317892,(((((TDEL0F00370_TORDC:...
3910  P19158_YEAST  (YALI0B07535g_YARLI:0.973069,(Q5A3W3_CANAL:1.8...
3911  P28519_YEAST  (O59753_SCHPO:0.835752,(YALI0B01474g_YARLI:1.1...
3912  P53723_YEAST  (O94280_SCHPO:1.03372,(Q59Z43_CANAL:0.664413,(...
3913  Q02648_YEAST  (O74516_SCHPO:1.04834,((((Q6CR42_KLULA:0.32224...
3914  P40207_YEAST  (Wican1_51760_HANAN:0.0076229,(((TDEL0G03290_T...
3915  P38989_YEAST  (P41003_SCHPO:0.380515,(((((Q6CQM4_KLULA:0.449...
3916  Q08227_YEAST  (Q9USQ6_SCHPO:0.966332,(((Q5AFL6_CANAL:1.02241...
3917  Q99297_YEAST  (Q09188_SCHPO:0.110083,((((Wican1_60885_HANAN:...
3918  Q12675_YEAST  (O36028_SCHPO:0.73681,(Q09891_SCHPO:0.896554,(...
3919  P30902_YEAST  (O94390_SCHPO:0.549936,((Q59PV8_CANAL:0.430312...
3920  P48526_YEAST  (O14160_SCHPO:0.858436,(((((Q59JY1_CANAL:0.287...
3921  P49166_YEAST  (P59289_SCHPO:0.0439648,((((Q6CUW0_KLULA:0.138...
3922  P36142_YEAST  (Q09906_SCHPO:0.929673,(YALI0B18260g_YARLI:1.0...
3923  P37838_YEAST  (O74400_SCHPO:1.3185,(((Q59KZ0_CANAL:0.402377,...
3924  P07244_YEAST  (P20772_SCHPO:0.357364,(((Wican1_29677_HANAN:0...
3925  P32585_YEAST  (YALI0B23320g_YARLI:0.621764,(Q59X40_CANAL:0.9...
3926  P19735_YEAST  (Q12381

In [29]:
#load GOIs here

## Build phylo ID conversion dataset

In [ ]:
#Make Dictionary for SC genes out of all_id_conversion.txt file
#Only need to do once
io_library.build_phylo_conversion_sc()

## Lookup functions

In [6]:
#check SC genename for given phylome name
gois = ['Phy000CZF5_YEAST', 'Phy000CX7L_YEAST', 'Phy000CZ7U_YEAST']
io_library.SC_common_name_lookup(list(sc_phylomedb_convert[sc_phylomedb_convert["phylomedb_name"].isin(gois)]['sc_genename']))

['SNZ1', 'SNZ2', 'SNZ3']

In [13]:
columns_to_view = ['SC_common_name_low','sc_genename_low', 'SC_common_name_high','sc_genename_high','log2FoldChange_SC_low', 'log2FoldChange_SC_high', 'log2FoldChange_KL_low','topologies_low','topologies_high','threshold_group_KL','threshold_group_SC']
sc_genename = ['YGR256W','YGR254W','YGR243W','YGR241C','YGR239C','YGR238C','YGR221C']
ohnologs_expression_sorted[(ohnologs_expression_sorted['sc_genename_low'].isin(sc_genename)) | (ohnologs_expression_sorted['sc_genename_high'].isin(sc_genename)) ].loc[:,columns_to_view]


SC_common_name_low sc_genename_low SC_common_name_high sc_genename_high  \
258               GND1         YHR183W                GND2          YGR256W   
259               ENO2         YHR174W                ENO1          YGR254W   
261               MPC2         YHR162W                MPC3          YGR243W   
262            YAP1802         YGR241C             YAP1801          YHR161C   
263              PEX21         YGR239C               PEX18          YHR160C   
264               KEL2         YGR238C                KEL1          YHR158C   
266               SKG6         YHR149C                TOS2          YGR221C   

     log2FoldChange_SC_low  log2FoldChange_SC_high  log2FoldChange_KL_low  \
258              -1.190240                7.143522               0.394234   
259              -0.975982                2.299404               0.559521   
261              -1.860001                3.387496              -1.789380   
262               0.062758                1.387725               1.183318   
263              -1.553162                5.889129              -1.657966   
264               0.050599                0.312086              -0.763434   
266              -2.719425               -2.151725              -2.633315   

     topologies_low  topologies_high threshold_group_KL threshold_group_SC  
258             NaN              NaN               flat            up_flat  
259             NaN              NaN               flat          flat_flat  
261             NaN              NaN               flat            up_flat  
262             NaN              NaN               flat          flat_flat  
263             NaN              NaN               flat            up_flat  
264             NaN              NaN               flat          flat_flat  
266             NaN              NaN               down          down_flat

In [142]:
gois = ['COX5A','COX5B']
print(io_library.SC_orf_lookup_by_name(gois))
trees[trees['sc_genename'].isin(io_library.SC_orf_lookup_by_name(gois))]

['YNL052W', 'YIL111W']


alignment_method    score  \
phylomedb_name                               
Phy000CZNA_YEAST               LG -6294.29   
Phy000CWSS_YEAST               LG -5039.33   

                                                               tree  \
phylomedb_name                                                        
Phy000CZNA_YEAST  (((((((((((Phy003LVZ0_51914:0.345273,Phy003M1S...   
Phy000CWSS_YEAST  ((((((((((Phy004FHA6_KAZAF:0.374036,((((Phy000...   

                 sc_genename  
phylomedb_name                
Phy000CZNA_YEAST     YNL052W  
Phy000CWSS_YEAST     YIL111W

In [144]:
#view tree for a given sc common name
#gois = ['ERV15','ERV14','AKR2','AKR1','YML018C','THI74','KDX1','SLT2','AAP1','APE2','MPC2','MPC3']
goi = ['COX5B']
tree = Tree(trees[trees['sc_genename']==io_library.SC_orf_lookup_by_name(goi)[0]]['tree'].values[0])
print(tree)

#prune down to seed and outgroup species
KLE_branch_seed = "KLULA"
ZT_branch_seed = "TORDC"
WGH_branch_seed = "YEAST" #options are "YEAST", "SACCA"
outgroups = ["CANAL","YARLI", "SCHPO"]   

leaf_names = tree.get_leaf_names()

# #Make a count of proteins from each species. 
# leaf_subset = [name for name in leaf_names if name.split('_')[1] in {WGH_branch_seed,KLE_branch_seed,ZT_branch_seed}]
# leaf_subset_spec = [name.split("_")[1] for name in leaf_subset]
# leaf_subset_spec_counter = Counter(leaf_subset_spec)

#Prune down tree to include only spec of interest and outgroups
leaf_subset_with_outgroups = [name for name in leaf_names if name.split('_')[1] in set([WGH_branch_seed,KLE_branch_seed,ZT_branch_seed] + outgroups)]
tree.prune(leaf_subset_with_outgroups) 
print(tree)

#Tree(trees[trees['sc_genename']==io_library.SC_orf_lookup_by_name(goi)[0]]['tree'])
#tree = Tree(trees.loc['Phy000CWKY_YEAST','tree'])


                              /-Phy004FHA6_KAZAF
                             |
                             |            /-Phy000NRSG_SACCA
                             |         /-|
                             |        |   \-Phy004F5JS_588726
                             |        |
                           /-|      /-|      /-Phy0023YX3_SACKL
                          |  |     |  |   /-|
                          |  |     |  |  |  |   /-Phy0000EKH_ASHGO
                          |  |     |   \-|   \-|
                          |  |   /-|     |      \-Phy0008MYC_KLULA
                          |  |  |  |     |
                          |  |  |  |      \-Phy002435W_LACTH
                        /-|   \-|  |
                       |  |     |   \-Phy004FVIW_TORDC
                       |  |     |
                       |  |     |   /-Phy000CZNA_YEAST
                       |  |      \-|
                     /-|  |         \-Phy000NPER_SACBA
                    |  |  |
               

['SNZ1', 'SNZ2', 'SNZ3'] - these genes have multiple orthologs in SC

## Single phylogenies for various combinations of seed species. 

In [ ]:
# Currently doesn't support different WGH seeds. 
#To use different WGH seed, we need to assign a related gene from the seed species for each S.Cer based tree, 
#and stop if that doesn't exist. 
#Run on the GPU cluster if you are doing a lot of branch seeds. 

## Should make the missing columns more specific. 

trees['topologies_{}_{}_{}'.format(WGH_branch_seed,KLE_branch_seed,ZT_branch_seed)] = topologies

KLE_branch_seeds = ["KLULA", "ASHGO", "LACTH","SACKL", "KLUWA"]  
ZT_branch_seeds = ["TORDC","ZYGRO"] 
#WGH_branch_seeds = ["YEAST","SACCA", "VANPO"]       
WGH_branch_seed = "YEAST"
#51660: N. Bacillosporus, 51914: C. CAstelii, K.Nag: 588726, N. Dairenensis: NAUDC, 
#1071379: Tet Blatt, TETPH: Tet. Phaffii, KAZAF: K. Africana, CANGA: Candida glabrata, SACBA
outgroups = ["CANAL","YARLI","DEKBR"]

for KLE_branch_seed in KLE_branch_seeds: 
    for ZT_branch_seed in ZT_branch_seeds:       
        print(" ".join([KLE_branch_seed, ZT_branch_seed, WGH_branch_seed]))
        topologies = io_library.assign_topologies(trees, WGH_branch_seed, KLE_branch_seed, ZT_branch_seed, outgroups)
        trees['topologies_{}_{}_{}'.format(WGH_branch_seed,KLE_branch_seed,ZT_branch_seed)] = topologies


trees.to_csv(data_processing_dir + os.path.normpath("phylogenetic_trees/trees_topologies.csv"))    


In [3]:
#Load tree_topology data
trees_topology = pd.read_csv(data_processing_dir + os.path.normpath('phylogenetic_trees/trees_topologies.csv'),index_col = 0)

In [7]:
topology_columns = [column for column in trees_topology.columns if column.split('_')[0]=='topologies']
a = Counter(trees_topology[topology_columns[0]])


5402

In [10]:
topology_column = topology_columns[0]

topology_counter = Counter(trees_topology[topology_column])

print(topology_counter)

#topologies1 = ['A', 'B', 'C', 'under_construction']



# 'missing KLULA protein'
# 'missing TORDC and KLULA protein'
# 'missing TORDC protein'

Counter({'A': 2714, 'C': 1216, 'missing KLULA protein': 479, 'missing TORDC and KLULA protein': 412, 'B': 251, 'under construction': 226, 'missing TORDC protein': 103, 'D': 1})


[479, 103, 412]

In [18]:
#Barplot of topology percentages for different seeds. 

topologies1 = ['A', 'B', 'C', 'under construction','D']
topologies2 = ['missing KLE protein', 'missing ZT protein', 'missing ZT and KLE protein']
topologies = topologies1 + topologies2

topology_columns = [column for column in trees_topology.columns if column.split('_')[0]=='topologies']

traces = [] 

for topology_column in topology_columns: 
    
    topology_counter = Counter(trees_topology[topology_column])
    
    topology_counts1 = [topology_counter[topology] for topology in topologies1]
    
    KLE_branch_seed = topology_column.split('_')[2]
    ZT_branch_seed = topology_column.split('_')[3]

    topologies2_dict = {'missing KLE protein': 'missing ' + KLE_branch_seed + ' protein', 
                        'missing ZT protein': 'missing ' + ZT_branch_seed + ' protein', 
                        'missing ZT and KLE protein' : 'missing ' + ZT_branch_seed + ' and ' + KLE_branch_seed + ' protein'}

    topology_counts2 = []
    for topology in topologies2:
        topology_counts2.append(topology_counter[topologies2_dict[topology]])

    topology_counts = topology_counts1 + topology_counts2
    total_counts = sum(topology_counts)
    if total_counts != sum(topology_counter.values()):
        print('Alert -- count mismatch!!!!')
        print(topology_counter)
        print(topology_counts)
    
    topology_pct = np.array(topology_counts)/total_counts
    
    trace = pygo.Bar(
        x= topologies,
        y= topology_pct,
        text= topology_counts,
        textposition = 'auto', 
        name = KLE_branch_seed + ' / ' + ZT_branch_seed 
#         marker=dict(
#             color='rgb(158,202,225)',
#             line=dict(
#                 color='rgb(8,48,107)',
#                 width=1.5),
#             ),
#         opacity=0.6
    )
    
    traces.append(trace)


py.iplot(traces, filename='topologies_single_gene')

High five! You successfully sent some data to your account on plotly. View your plot in your browser at https://plot.ly/~heineike02_student/0 or inside your plot.ly account where it is named 'topologies_single_gene'


In [56]:
#Visualize with a heatmap

ohnologs_topology_sorted = ohnologs_topology.sort_values(by = ['threshold_group_SC','log2FoldChange_KL_low'])

#fold change values
fc_values = ohnologs_topology_sorted.loc[:,['log2FoldChange_SC_low', 'log2FoldChange_SC_high', 'log2FoldChange_KL_low']]


#Topology values
WGH_branch_seed = 'YEAST'

topology_values = []
topology_column_labels = []

for fc_group in ['low', 'high']:
    for topology_column in topology_columns: 
        print(fc_group + " " + topology_column)
        KLE_branch_seed = topology_column.split('_')[2]
        ZT_branch_seed = topology_column.split('_')[3]
        value_dict = {'A':-5, 'B':5,'C':0,
                      'under construction':3,
                      'missing ' + KLE_branch_seed + ' protein':-1,
                      'missing ' + ZT_branch_seed + ' and ' + KLE_branch_seed +' protein':-2,
                      'missing ' + ZT_branch_seed + ' protein':-1.5,
                       np.nan:-2.5
                     }
        topology_values.append([value_dict[topology] for topology in ohnologs_topology_sorted[topology_column + '_' + fc_group]])
        
        topology_column_labels = WGH_branch_seed[0] + KLE_branch_seed[0] + ZT_branch_seed[0] + '_' + fc_group


gene_labels = ohnologs_topology_sorted['SC_common_name_low']+ ' / ' + ohnologs_topology_sorted['SC_common_name_high']

trace1 = pygo.Heatmap(z = np.array(fc_values),
                   x=['SC low FC', 'SC high FC', 'KL FC'],
                   y= gene_labels)
trace2 = pygo.Heatmap(z = np.array(topology_values).transpose(),
                   x=topology_column_labels,
                   y= gene_labels)


fig = pytools.make_subplots(rows=1, cols=2)

fig.append_trace(trace1, 1, 1)
fig.append_trace(trace2, 1, 2)

#fig['layout'].update(height=600, width=600, title='i <3 subplots')

# layout = pygo.Layout(
#     yaxis = {'title': 'Low Paralog / High Paralog'}
# )

#fig = pygo.Figure(data=data, layout=layout)
py.iplot(fig, filename='topology_heatmap')




low topologies_YEAST_KLULA_TORDC
low topologies_YEAST_KLULA_ZYGRO
low topologies_YEAST_ASHGO_TORDC
low topologies_YEAST_ASHGO_ZYGRO
low topologies_YEAST_LACTH_TORDC
low topologies_YEAST_LACTH_ZYGRO
low topologies_YEAST_SACKL_TORDC
low topologies_YEAST_SACKL_ZYGRO
low topologies_YEAST_KLUWA_TORDC
low topologies_YEAST_KLUWA_ZYGRO
high topologies_YEAST_KLULA_TORDC
high topologies_YEAST_KLULA_ZYGRO
high topologies_YEAST_ASHGO_TORDC
high topologies_YEAST_ASHGO_ZYGRO
high topologies_YEAST_LACTH_TORDC
high topologies_YEAST_LACTH_ZYGRO
high topologies_YEAST_SACKL_TORDC
high topologies_YEAST_SACKL_ZYGRO
high topologies_YEAST_KLUWA_TORDC
high topologies_YEAST_KLUWA_ZYGRO
This is the format of your plot grid:
[ (1,1) x1,y1 ]  [ (1,2) x2,y2 ]



## Ortholog Pair Topology

In [26]:
#For ortholog pairs, find topology 

ortholog_dataset = ortholog_dataset_goi

#ortholog_dataset = ohnologs_expression_sorted.loc[:,['sc_genename_low','sc_genename_high']].rename(columns = {'sc_genename_low': 'low','sc_genename_high': 'high'})
#ohnologs_expression_sorted.iloc[0:30,:]
#ohnologs_expression_sorted[ohnologs_expression_sorted['SC_common_name_low'].isin(goi_sc_names)]  

#Input a series which has two columns - one for each paralog.  The labels distinguish the two paralogs. 

KLE_branch_seed = "KLULA"
ZT_branch_seed = "TORDC"
WGH_branch_seed = "YEAST" #options are "YEAST", "SACCA"
outgroups = ["CANAL","YARLI", "SCHPO"]    
verbose = False

#Input: 
#trees:  Dictionary which is indexed on phylome db name and hass the tree, sc_genename, alignment_method, and score

##
#sc_phylomedb_convert:  Should load this with the function. 

## Begin function

#Load sc_phylomedb_convert data
sc_phylomedb_convert = pd.read_table(data_processing_dir + os.path.normpath("phylogenetic_trees/sc_id_conversion.txt"), header=None)
sc_phylomedb_convert.columns = ["phylomedb_name","sc_genename"]

#The ortholog dataset has a column with two pairs of orthologs.  This argument is the label that indicates the group that the orthologs come from.
#Typically I use the fold change levels 'low' and 'high'
paralog_labels = list(ortholog_dataset.columns)  #indicates low or high fold change in a sorted expression dataset. 
#dictionary to identify opposite label.
other_label = {paralog_labels[0]:paralog_labels[1], paralog_labels[1]:paralog_labels[0]}

#record when a paralog pair has a missing paralog: 
missing_paralog_in_tree = []
#records single gene topologies for each paralog
topologies = []
#records joint topology for a paralog pair. 
joint_topologies = []

for ind, row in ortholog_dataset.iterrows():
    sc_genes = {label: row[label] for label in paralog_labels}
    
    #Attempts to assign a row from the trees data
    tree_rows = {label: trees[trees['sc_genename']==sc_genes[label]] for label in paralog_labels}
        
    #assign phylome db gene name for each label. 
    genes = {}
    for label in paralog_labels:
        if len(tree_rows[label])==0: 
            #If sc_genename not linked to a row in the tree database, then either
            #A: there is no name in phylome db for the gene of interest or 
            #B: thre is no tree in the given trees database for that gene.
            #either way assigns None to tree_rows
            tree_rows[label] = None
            print("no tree for ortholog " + sc_genes[label] + "in ortholog pair " + sc_genes[paralog_labels[0]] + " " + sc_genes[paralog_labels[1]])
            phylome_db_lookup_name = sc_phylomedb_convert[sc_phylomedb_convert['phylomedb_name']==sc_genes[label]]
            if len(phylome_db_lookup_name)==0: 
                #Case A
                print("no name in phylome db for " + sc_genes[label])
                genes[label]= None
            else:
                #Case B
                genes[label]= phylome_db_lookup_name['phylomedb_name'].values[0]
                print("there is a gene name in phylome db: " + genes[label])
        else: 
            genes[label] = tree_rows[label].index.values[0]     

    #skip to next set of paralogs if either of them don't have a tree_row. 
    if any([item is None for key,item in tree_rows.items()]): 
        #joint topology is not possible if one paralog is not present. 
        none_none = {label: None for label in paralog_labels}
        joint_topology = none_none
        
        #boolean which indicates which values have tree rows - remember one will be None
        tree_row_absent = [item is None for item in list(tree_rows.values())]
        tree_row_present = [not item for item in tree_row_absent]
        
        if all(tree_row_absent):
            #if both tree rows are not present
            print("Neither paralog had a tree " + sc_genes[paralog_labels[0]] + " " + sc_genes[paralog_labels[1]])
            topology = {label: none_none for label in paralog_labels}
        else: 
            ##Should actually assign correct single topology if only one is ok. 
            
            #find label with tree
            for ind,label in enumerate(list(tree_rows.keys())):
                if tree_row_present[ind]:
                    print(label)
                    label_with_tree = label
                    print(other_label[label])
                    label_wo_tree = other_label[label]
            
            print(sc_genes[label_with_tree] + ' had a tree but its paralog, ' + sc_genes[label_wo_tree] + ' did not.')
            
            #label with tree should have a single topology for its tree but no topology for the other tree
            topology_label_with_tree = {}
            
            ## To do here

            #calculate topology for 
#             tree = Tree(tree_string)
#             #Extract all leaves that are are WGH_branch_seed, KLE_branch_seed, ZT_branch seed 
#             leaf_names = tree.get_leaf_names()
#             leaf_subset = [name for name in leaf_names if name.split('_')[1] in {WGH_branch_seed,KLE_branch_seed,ZT_branch_seed}]
            
#             topology_label_with_tree[label_with_tree] = io_library.assign_single_topology(gene,leaf_subset_label,tree, KLE_branch_seed, WHG_branch_seed, ZT_branch_seed, verbose)
            topology_label_with_tree[label_with_tree] = 'should calculate single topology'
            topology_label_with_tree[label_wo_tree] = None
            topology[label_with_tree] = topology_label_with_tree 
            
            #label without tree should have no topology for its tree but might have a topology for the other tree
            topology_label_wo_tree = {}
            topology_label_wo_tree[label_wo_tree] = None
            topology_label_wo_tree[label_with_tree] = 'should check if gene name is present, if so if it is in the tree, and if so calculate the single topology'
        topologies.append(topology)
        joint_topologies.append(joint_topology)
        continue

    #Calculate joint topology and topology for paralog pair when both paralogs have a tree. 
    
    tree_strings = {label: tree_rows[label]['tree'][0] for label in tree_rows.keys()}
   
    joint_topology = {}
    topology = {}
    #iterate through trees for both paralogs
    for label_tree, tree_string in tree_strings.items():
        #Test string "(Phy000D0CO_YEAST:0.0377129,Phy000NQ6C_SACBA:0.0653462,((Phy004FKRY_NAUDC:0.0912444,Phy000NS92_SACCA:0.115243)0.99985:0.0765755,((((Phy004F4O4_588726:0.253337,(Phy004F7XM_1071379:0.214507,(Phy000JNPU_VANPO:0.153757,Phy004FCCE_TETPH:0.171821)0.99985:0.111541)0.998964:0.0332526)0.980573:0.0259834,(((Phy000JRAI_KLUWA:0.065763,Phy002423J_LACTH:0.0853684)0.99985:0.118867,(Phy000NWPH_SACKL:0.118849,(Phy0000C6I_ASHGO:0.163073,Phy0008PPH_KLULA:0.187974)0.999239:0.0291304)0.99985:0.036076)0.99985:0.0367641,(Phy004FVSC_TORDC:0.14757,Phy00244CV_ZYGRO:0.280471)0.99985:0.0455249)0.99985:0.0234786)0.952493:0.0167291,(Phy004FGQ7_KAZAF:0.157457,(Phy00042WG_CANGA:0.179752,(Phy003M12E_51660:0.219913,Phy003LT0W_51914:0.298372)0.99985:0.0535978)0.99985:0.0516007)0.953575:0.0196913)0.972812:0.0235719,((Phy000NM4H_SACBA:0.0797609,Phy000CVUB_YEAST:0.111285)0.99985:0.241968,(((Phy004G2AN_HANAN:0.607512,((Phy0043HAA_DEKBR:0.613043,(Phy0002J8L_CANAL:0.261718,(Phy000M23V_PICST:0.182477,Phy0005LWI_DEBHA:0.179178)0.99985:0.0567132)0.99985:0.131607)0.99985:0.0941789,(Phy000EWFJ_YARLI:0.646307,((Phy004D9WH_45786:0.295069,Phy004DAUY_45786:0.786718)0.99985:0.125919,(Phy000D148_SCHPO:0.930355,((Phy004D9WR_45786:0.198548,((Phy004G3IF_HANAN:0.174478,(Phy0000C5V_ASHGO:0.212532,(((Phy004F8NL_1071379:0.141367,(Phy004FD9R_TETPH:0.141431,Phy000JLBS_VANPO:0.116295)0.99985:0.0650686)0.99985:0.0356093,((((Phy004F1VH_588726:0.164592,(Phy000NUHY_SACCA:0.0565971,Phy004FI75_NAUDC:0.0592237)0.99985:0.032963)0.967128:0.0146652,(Phy003M2DA_51660:0.0718156,Phy003LSK0_51914:0.123436)0.99985:0.0352864)0.993589:0.0159514,(Phy000CY5Z_YEAST:0.0752616,(Phy00042VW_CANGA:0.0982544,(Phy004FH3D_KAZAF:0.0433095,Phy004FGT3_KAZAF:0.0291784)0.99985:0.0655249)0.998957:0.0272474)0.991005:0.0227564)0.99985:0.0179636,(Phy004FYCC_TORDC:0.0799393,Phy00244D9_ZYGRO:0.119371)0.994601:0.0197164)0.99985:0.0232753)0.99985:0.060257,(Phy0023ZJ5_SACKL:0.0790576,(Phy0008M3B_KLULA:0.284,(Phy002433O_LACTH:0.034923,Phy000JPRE_KLUWA:0.0533176)0.99985:0.0827682)0.99985:0.0294055)0.934264:0.0162554)0.99985:0.0331771)0.99985:0.0953258)0.99985:0.104003,(Phy0043EA7_DEKBR:0.329095,(Phy0002IUY_CANAL:0.113784,(Phy0005HV2_DEBHA:0.168247,Phy000M1Q7_PICST:0.0681382)0.2608:0.0373838)0.99985:0.142364)0.949433:0.0296014)0.99985:0.058341)0.99985:0.439321,((Phy000D147_SCHPO:0.337686,(Phy0043DGZ_DEKBR:0.179192,((Phy000NXUS_SACKL:0.0119718,(Phy000JQ10_KLUWA:0.00356342,Phy00240LJ_LACTH:0.0173874)0.99985:0.0420066)0.973509:0.0124443,(Phy000PHQX_KLULA:0.0837091,((Phy004FW12_TORDC:0.0270097,Phy00245C9_ZYGRO:0.0799646)0.99985:0.0293122,((Phy003LVON_51914:0.00488701,Phy003LVW9_51914:0.0214526)0.99985:0.049813,((Phy004FB62_TETPH:0.0450363,Phy000JO2T_VANPO:0.0157048)0.99985:0.0284698,((Phy004F55K_588726:0.043957,((Phy003ET24_SACBA:9.8e-08,Phy003ESPQ_SACBA:0.00517011)0.99985:0.0245326,(Phy000CZIO_YEAST:0.00368032,Phy000CW40_YEAST:0.00684232)0.99985:0.0147427)0.99985:0.0233092)0.99985:0.0129149,(((Phy004FL32_NAUDC:8.89e-08,Phy004FJ8W_NAUDC:0.0210664)0.99985:0.018709,(Phy000NR38_SACCA:0.00441582,Phy000NRWI_SACCA:0.0338758)0.99985:0.0200783)0.848919:0.00637556,(Phy000HL5M_CANGA:0.0380492,(Phy004FFGC_KAZAF:0.00245352,Phy004FHDM_KAZAF:0.00791846)0.99985:0.0383082)0.99985:0.0206773)0.929977:0.010802)0.99985:0.0134442)0.98884:0.0114619)0.997822:0.00886798)0.99985:0.0274939)0.99985:0.0182694)0.99985:0.146112)0.99985:0.27075)0.99985:0.616952,((Phy000NXBP_SACKL:0.0276937,((Phy0000DBL_ASHGO:0.0543494,(Phy000JPNA_KLUWA:0.0151433,Phy00243RQ_LACTH:0.00665339)0.99985:0.0350212)0.99985:0.02568,(Phy0008O77_KLULA:0.0552353,((Phy004FCXO_TETPH:0.0382825,(Phy000JO2F_VANPO:0.030324,Phy000JND8_VANPO:0.0667573)0.99985:0.0551445)0.99985:0.0188317,(Phy004F80P_1071379:0.0787646,((Phy004F1YX_588726:0.0731204,(Phy004FFWL_KAZAF:0.0212898,((Phy000CYMB_YEAST:0.0105865,Phy000NLRW_YEAST:0.0212602)0.99985:0.0105349,(Phy003ESNK_SACBA:0.0190025,Phy000NLWS_SACBA:0.0299389)0.99985:0.00986892)0.99985:0.0227824)0.97486:0.00873849)0.970244:0.00473995,((Phy000457A_CANGA:0.0545661,(Phy004FKDA_NAUDC:0.0379504,Phy000NR20_SACCA:0.0526524)0.953743:0.014551)0.994092:0.0148149,((Phy003LUN2_51914:0.0422204,Phy003M1LP_51660:0.036874)0.995219:0.011347,(Phy004FVH1_TORDC:0.0373219,Phy00244HG_ZYGRO:0.100079)0.99985:0.0412169)0.921431:0.00903586)0.998632:0.0123235)0.99985:0.0134243)0.985352:0.00847894)0.99985:0.0739419)0.99985:0.0195691)0.998527:0.0148528)0.99985:0.0749685,(Phy004G2XL_HANAN:0.14119,(((Phy000EWZ3_YARLI:0.0811337,(Phy000EUZU_YARLI:0.0243446,(Phy003G9YQ_YARLI:0.0484158,Phy000EXEP_YARLI:0.0328251)0.995874:0.0170054)0.99985:0.0245836)0.99985:0.0600747,(Phy004D8CO_45786:0.151996,(Phy004DBR1_45786:0.0452565,Phy004DANJ_45786:0.0687919)0.99985:0.028464)0.99985:0.0358608)0.99985:0.0642867,((Phy0002HWT_CANAL:0.0628017,(Phy0005M0G_DEBHA:0.102678,Phy000M2EQ_PICST:0.0437075)0.99985:0.0338248)0.99985:0.0940611,((Phy0002LM8_CANAL:0.0839271,Phy0005M8P_DEBHA:0.180701)0.99985:0.0500089,(((Phy001SWG1_SCHPO:0.0352369,Phy001SWG2_SCHPO:0.063444)0.99985:0.199495,(Phy0043E6V_DEKBR:0.103961,Phy0043H82_DEKBR:0.133998)0.99985:0.110704)0.99985:0.042867,(Phy004G3G7_HANAN:0.0982853,((Phy00240WA_LACTH:0.0451633,Phy000JR1W_KLUWA:0.0681361)0.99985:0.0587049,((Phy003FKGG_ASHGO:0.0811873,(Phy000NYEU_SACKL:0.0536055,Phy0008N6P_KLULA:0.0923059)0.77818:0.00986821)0.998707:0.0219914,(Phy004FVHE_TORDC:0.063156,(((Phy000NU01_SACCA:0.0890922,(((Phy004F8XZ_1071379:0.132533,(Phy003LSW3_51914:0.107247,Phy003M3EY_51660:0.0677824)0.99985:0.0412051)0.858533:0.00671357,(Phy004FKS7_NAUDC:0.0665065,(Phy000JLXL_VANPO:0.0625492,(Phy004FCBV_TETPH:0.145474,Phy000JNJG_VANPO:0.0837134)0.99985:0.034268)0.99985:0.0430787)0.998837:0.0208369)0.99985:0.0196212,(Phy000457W_CANGA:0.0711854,(Phy004F2YU_588726:0.0564193,(Phy003ET83_SACBA:0.0412039,Phy000CX4H_YEAST:0.0312003)0.99985:0.0837373)0.947318:0.0223552)0.941307:0.00984725)0.99985:0.011665)0.99985:0.0192451,(Phy004FH1A_KAZAF:0.0907382,(Phy000CVMU_YEAST:0.0230885,Phy000NM19_SACBA:0.0378338)0.99985:0.077243)0.993676:0.0240291)0.99985:0.0203553,(Phy00244KO_ZYGRO:0.227036,(Phy000NT8F_SACCA:0.0706802,Phy004FLUW_NAUDC:0.102082)0.99985:0.0400108)0.951593:0.0156376)0.821965:0.0142286)0.99985:0.0555068)0.995297:0.0215114)0.99985:0.0617542)0.99985:0.0766563)0.99985:0.0391873)0.99985:0.0512995)0.868152:0.00804878)0.986886:0.0281588)0.895443:0.0194272)0.99985:0.201861)0.987571:0.16679)0.99985:3.18348)0.99985:0.571344)0.871722:0.0680403)0.99985:0.176123)0.99985:0.0800853)0.99985:0.182843,(Phy004F7V6_1071379:1.13573,(Phy003M25Z_51660:1.38478,Phy003LWAK_51914:1.88325)0.99985:0.607834)0.985457:0.149056)0.99985:0.0802184,((Phy004FDRT_TETPH:0.413917,Phy000JMLE_VANPO:0.43574)0.99985:0.102264,(Phy004FLX8_NAUDC:0.23091,Phy000NSTN_SACCA:0.23412)0.99985:0.102889)0.337602:0.0337093)0.99985:0.0474337)0.924651:0.0456035)0.99985:0.0507224)0.99985:0.102691);"
        tree = Tree(tree_string)

        #Extract all leaves that are are WGH_branch_seed, KLE_branch_seed, ZT_branch seed 
        leaf_names = tree.get_leaf_names()
        leaf_subset = [name for name in leaf_names if name.split('_')[1] in {WGH_branch_seed,KLE_branch_seed,ZT_branch_seed}]

        #Check that both WGH proteins are present
        if len(set(genes.values())-set(leaf_subset)) > 0: 
            missing_paralog_in_tree.append(sc_genes)
            print("both paralogs not in tree: {} {}".format(*sc_genes.values()))
            joint_topology[label_tree] = None
            ##should find topology for just one that exists
            print("could find topology just for the one that exists, but for now just assigning tree_missing_paralog")
            topology[label_tree] = "tree_missing_paralog"
            continue
            
  
        #Make a count of proteins from each species. 
        leaf_subset_spec = [name.split("_")[1] for name in leaf_subset]
        leaf_subset_spec_counter = Counter(leaf_subset_spec)
        
        
        #Prune down tree to include only spec of interest and outgroups
        leaf_subset_with_outgroups = [name for name in leaf_names if name.split('_')[1] in set([WGH_branch_seed,KLE_branch_seed,ZT_branch_seed] + outgroups)]
        tree.prune(leaf_subset_with_outgroups) 

        
        #Throw out extra ZT and KLE branch gene if children of common ancestor between it and current gene are in an outgroup.  
        for branch_seed in [KLE_branch_seed,ZT_branch_seed]:
            if (leaf_subset_spec_counter[branch_seed]>1):
                #If there is more than one gene from the current branch 
                
                #for each gene from the current branch: 
                branch_genes = [leaf for ind,leaf in enumerate(leaf_subset) if leaf_subset_spec[ind]==branch_seed]
                for branch_gene in branch_genes:
                    #get common ancestor between both paralogs and branch gene
                    ancestor = tree.get_common_ancestor([genes[paralog_labels[0]],genes[paralog_labels[1]],branch_gene])
                    ancestor_leaves = ancestor.get_leaf_names()
                    #find species of leaves of common ancestor
                    ancestor_leaves_spec = [name.split("_")[1] for name in ancestor_leaves]

                    #if the leaves of the parent node include an outgroup, throw the gene out of the leaf subset and update counter
                    if len(set(ancestor_leaves_spec) & set(outgroups)) > 0:
                        if verbose: 
                            print('Threw out ' + branch_gene + ' because outgroup gene between it and seed')
                        leaf_subset.remove(branch_gene)    
                        leaf_subset_spec = [name.split("_")[1] for name in leaf_subset]
                        leaf_subset_spec_counter = Counter(leaf_subset_spec)

                    if  (leaf_subset_spec_counter[branch_seed]==0):
                        if verbose: 
                            print('Alert threw out both ' + branch_seed + ' genes for seed: ' + genes[paralog_labels[0]] + ' ' + genes[paralog_labels[1]])

        #If there are extra WGH branch genes, throw them out.  Keep track of whether the extra genes had an outgroup between paralogs or not.  
        if (leaf_subset_spec_counter[WGH_branch_seed]>2):
           
            #get extra WGH genes
            not_WGH = [spec != WGH_branch_seed for spec in leaf_subset_spec]
            is_paralog = [leaf in set(genes.values()) for leaf in leaf_subset]
            is_other_WGH = [not(ig or ns) for ig,ns in zip(is_paralog, not_WGH)]
            other_WGH = [gene for gene,cond in zip(leaf_subset,is_other_WGH) if cond]
            
            #for each of extra WGH genes. 
            for other_WGH_gene in other_WGH:
                ancestor = tree.get_common_ancestor([genes[paralog_labels[0]],genes[paralog_labels[1]],other_WGH_gene])
                ancestor_leaves = ancestor.get_leaf_names()
                ancestor_leaves_spec = [name.split("_")[1] for name in ancestor_leaves]

                #if the leaves of the parent node include an outgroup, throw the gene out of the leaf subset and update counter
                if len(set(ancestor_leaves_spec) & set(outgroups)) > 0:
                    if verbose: 
                        print('Threw out ' + other_WGH_gene + ' because outgroup gene between it and seed')
                    leaf_subset.remove(other_WGH_gene)    
                    leaf_subset_spec = [name.split("_")[1] for name in leaf_subset]
                    leaf_subset_spec_counter = Counter(leaf_subset_spec)
            
#             #why do I need to remove these? 
#             if leaf_subset_spec_counter[WGH_branch_seed]>2: 
#                 if verbose:
#                     print('After checking for outgroups, still have more than 2 ' + WGH_branch_seed + ' Genes for ' 
#                           + genes[paralog_labels[0]] + ' ' + genes[paralog_labels[1]] + '.  Removed extras.')
#                 not_WGH = [spec != WGH_branch_seed for spec in leaf_subset_spec]
#                 is_paralog = [leaf in set(genes.values()) for leaf in leaf_subset]
#                 is_other_WGH = [not(ig or ns) for ig,ns in zip(is_paralog, not_WGH)]
#                 other_WGH = [gene for gene,cond in zip(leaf_subset,is_other_WGH) if cond]
#                 for other_WGH_gene in other_WGH:
#                     leaf_subset.remove(other_WGH_gene)
#                     leaf_subset_spec = [name.split("_")[1] for name in leaf_subset]
#                     leaf_subset_spec_counter = Counter(leaf_subset_spec)
        
        
        #Check to see if outgroup exists between two paralogs. 
                
        topology_tree = {}
        tree_outgroup_test = tree.copy()
        ancestor = tree_outgroup_test.get_common_ancestor([genes[paralog_labels[0]],genes[paralog_labels[1]]])
        ancestor_leaves = ancestor.get_leaf_names()
        ancestor_leaves_spec = [name.split("_")[1] for name in ancestor_leaves]
        ancestor_leaves_outgroup = [leaf for ind,leaf in enumerate(ancestor_leaves) if ancestor_leaves_spec[ind] in set(outgroups)]
        
        if len(set(outgroups) & set(ancestor_leaves_spec))>0:
            print("OG Pair: {} and {} : outgroup present between paralogs.".format(genes[paralog_labels[0]],genes[paralog_labels[1]]))
            
            #Assign joint topology
            joint_topology[label_tree] = 'C-outgroup-?'
            
            #Determine single topology - if it is the one outside, topology is "C-outgroup" if
            #If it is the one inside do a normal topology determination. 
            
            #high_low_swap = {'high':'low', 'low':'high'}
            for label, gene in genes.items(): 
                other_paralog = genes[other_label[label]]
                ancestor_gene = tree_outgroup_test.get_common_ancestor([gene] + ancestor_leaves_outgroup)
                ancestor_gene_leaves = ancestor_gene.get_leaf_names()
                if other_paralog in set(ancestor_gene_leaves):
                    topology_tree[label] = "C-outgroup"
                
                else: 
                    #Keeps all ZT and KLE branch leaves as well as the one for the current gene. 
                    leaf_subset_label = [leaf for ind, leaf in enumerate(leaf_subset) if (leaf_subset_spec[ind]!=WGH_branch_seed) | (leaf == gene) ]
                    topology_tree[label] = io_library.assign_single_topology(gene,leaf_subset_label,tree, KLE_branch_seed, WGH_branch_seed, ZT_branch_seed, verbose)

                topology[label_tree] = topology_tree   

            continue
        
        #prune tree for the rest of the calculation - keep only the leaf_subset and no outgroups.  
        tree.prune(leaf_subset)
               
        for label, gene in genes.items(): 
            #Keeps all ZT and KLE branch leaves as well as the one for the current gene.
            leaf_subset_gene = [leaf for ind, leaf in enumerate(leaf_subset) if (leaf_subset_spec[ind]!=WGH_branch_seed) | (leaf == gene) ]
            topology_tree[label] = io_library.assign_single_topology(gene,leaf_subset_gene,tree, KLE_branch_seed, WGH_branch_seed, ZT_branch_seed, verbose)
            
        #use topology of both and relationship between two paralogs to decide fine resolution topology
        joint_topology_out, topology_tree_updated = io_library.assign_joint_topology(genes, leaf_subset, tree, topology_tree, paralog_labels, sc_genes, KLE_branch_seed, WGH_branch_seed, ZT_branch_seed, verbose)
        joint_topology[label_tree] = joint_topology_out
        topology[label_tree] = topology_tree_updated
    
    topologies.append(topology)
    joint_topologies.append(joint_topology)

# #[io_library.SC_common_name_lookup(list(paralog_pair.values())) for paralog_pair in missing_paralog_in_tree]



both paralogs not in tree: YJL137C YKR058W
could find topology just for the one that exists, but for now just assigning tree_missing_paralog
both paralogs not in tree: YOL024W YFR017C
could find topology just for the one that exists, but for now just assigning tree_missing_paralog
both paralogs not in tree: YOL024W YFR017C
could find topology just for the one that exists, but for now just assigning tree_missing_paralog
OG Pair: Phy000CY89_YEAST and Phy000CWZL_YEAST : outgroup present between paralogs.
OG Pair: Phy000CY89_YEAST and Phy000CWZL_YEAST : outgroup present between paralogs.
both paralogs not in tree: YNL053W YIL113W
could find topology just for the one that exists, but for now just assigning tree_missing_paralog
OG Pair: Phy000CZN9_YEAST and Phy000CWSQ_YEAST : outgroup present between paralogs.
OG Pair: Phy000CYDX_YEAST and Phy000CZ8H_YEAST : outgroup present between paralogs.
OG Pair: Phy000CYDX_YEAST and Phy000CZ8H_YEAST : outgroup present between paralogs.
OG Pair: Phy000C

In [97]:
len(topologies)

65

These remarks relate to an earlier run. 

MBR1 didn't have a tree but ISF1 did.  Strange
tree not present for ortholog pair Phy000CZ7E_YEAST Phy000CZ83_YEAST
tree not present for ortholog pair Phy000CXBK_YEAST Phy000CXB5_YEAST

tree not present for ortholog pair Phy000CZAI_YEAST Phy000CXFP_YEAST

After checking for outgroups, still have more than 2 YEAST Genes.Phy000CZ3A_YEAST Phy000CYZW_YEAST
IMD3 and IMD4 are the proteins but IMD2 shows up as similar to IMD3W (and closer than IMD4)  
IMD1, IMD2 (Phy000CY0B_YEAST), YAR075W  - I wonder if that third one had any expression. 


After checking for outgroups, still have more than 2 YEAST Genes.Phy000CWKY_YEAST Phy000CXWI_YEAST
These are the hexose transporters which are extra duplicated in S.Cer: 
['Phy000CX94_YEAST', 'Phy000CZ5A_YEAST', 'Phy001SPOB_YEAST', 'Phy000CWKV_YEAST', 'Phy000CWKW_YEAST', 'Phy000CY0K_YEAST', 'Phy000CZRB_YEAST','Phy000CXWG_YEAST', 'Phy000CWKY_YEAST','Phy000CXWI_YEAST',  'Phy000CY0P_YEAST', 'Phy000CYP9_YEAST']
['HXT11',
 'HXT7',
 'HXT6',
 'HXT4',
 'HXT8',
 'HXT9',
 'HXT3',
 'GAL2',
 'HXT2',
 'HXT5',
 'HXT1',
 'HXT10']

When both TORDC and both KLULA genes often get thrown out for the same seed - maybe there is a wierd outgroup gene causing that? 

missing ortholog in tree: 
[['GLG2', 'GLG1'],
 ['YOL024W', 'IGD1'],
 ['YOL024W', 'IGD1'],
 ['MSG5', 'SDP1'],
 ['PIG1', 'GAC1'],
 ['SKT5', 'SHC1']]

Genes.Phy000CZ3A_YEAST Phy000CYZW_YEAST

## Integrate and visualise joint topology data. 

In [31]:
#Integrate data into goi_data: 

#Add data to input ortholog dataset
paralog_labels = ['low','high']
for tree_label in paralog_labels:
    joint_topologies_tree_label = [joint_topology[tree_label] for joint_topology in joint_topologies]
    ortholog_dataset_goi['joint_topologies_'+tree_label+'_tree'] = joint_topologies_tree_label
    
    for label in paralog_labels: 
        topologies_label_tree_label = []
        for topology in topologies: 
            if isinstance(topology[tree_label],dict):
                topologies_label_tree_label.append(topology[tree_label][label])
            elif isinstance(topology[tree_label],str):
                topologies_label_tree_label.append(topology[tree_label])
        ortholog_dataset_goi['topologies_' + tree_label + '_tree_' + label] = topologies_label_tree_label

#stack new data and change 'low'/'high' combined column to 'sc_genename' for merger
ortholog_dataset_goi_stack = pd.DataFrame(columns = ['sc_genename','joint_topologies_low_tree','joint_topologies_high_tree','topologies_low_tree', 'topologies_high_tree'])
for label in paralog_labels: 
    data_to_stack = ortholog_dataset_goi.loc[:,[label,'joint_topologies_low_tree','joint_topologies_high_tree','topologies_low_tree_' + label, 'topologies_high_tree_' + label]]
    data_to_stack.columns = ['sc_genename','joint_topologies_low_tree','joint_topologies_high_tree','topologies_low_tree', 'topologies_high_tree']
    ortholog_dataset_goi_stack = ortholog_dataset_goi_stack.append(data_to_stack)
    

goi_data_phylogenies = goi_data.merge(ortholog_dataset_goi_stack, how = 'left', on = 'sc_genename')

#add phylomedb_name column
phylomedb_name = io_library.get_phylomedb_name(list(goi_data_phylogenies['sc_genename']))
goi_data_phylogenies['phylomedb_name'] = phylomedb_name

goi_data_phylogenies.to_csv(data_processing_dir + os.sep + "20170817_klscpka" + os.sep + "goi_data_phylogenetic_trees.csv")


Unnamed: 0 SC_common_name sc_genename  log2FoldChange_SC  \
0          3968           GLK1     YCL040W           4.160390   
1          3969           EMI2     YDR516C           2.464130   
2          3972           SKT5     YBL061C          -1.806729   
3          3973           SHC1     YER096W           3.793451   
4          3978           GND2     YGR256W           7.143522   
5          3979           GND1     YHR183W          -1.190240   
6          3994           LPD1     YFL018C           1.016738   
7          3995          IRC15     YPL017C           3.049381   
8          4018           ENO1     YGR254W           2.299404   
9          4019           ENO2     YHR174W          -0.975982   
10         4118           SDH3     YKL141W           1.507281   
11         4119           SHH3     YMR118C           6.163424   
12         4144        YBR238C     YBR238C          -3.219891   
13         4145           RMD9     YGL107C          -0.048634   
14         4148           NDE2     YDL085W           7.299282   
15         4149           NDE1     YMR145C          -1.401044   
16         4154           HXT5     YHR096C           7.987038   
17         4155           HXT3     YDR345C          -2.024423   
18         4158           URA5     YML106W          -2.205157   
19         4159          URA10     YMR271C           4.621205   
20         4210           GIP2     YER054C           5.784015   
21         4211           PIG2     YIL045W           1.843841   
22         4310          COX5A     YNL052W           0.228952   
23         4311          COX5B     YIL111W           3.206427   
24         4350           SOL4     YGR248W           7.194901   
25         4351           SOL3     YHR163W          -0.150099   
26         4364           CIT2     YCR005C          -2.464709   
27         4365           CIT1     YNR001C           2.403591   
28         4368           HXK1     YFR053C           5.119539   
29         4369           HXK2     YGL253W          -3.237243   
..          ...            ...         ...                ...   
99         4656           MSN4     YKL062W           1.098421   
100        4657           MSN2     YMR037C           0.497774   
101        4676           EDS1     YBR033W           2.830364   
102        4677           RGT1     YKL038W           1.061525   
103        4734           TOD6     YBL054W          -4.042864   
104        4735           DOT6     YER088C           1.223284   
105        3966           NTH2     YBR001C           2.869520   
106        3967           NTH1     YDR001C           2.815195   
107        4072           GSY1     YFR015C           4.666016   
108        4073           GSY2     YLR258W           4.175972   
109        4074         RPL26B     YGR034W          -4.431738   
110        4075         RPL26A     YLR344W          -4.121747   
111        4204          RPS1A     YLR441C          -3.483954   
112        4205          RPS1B     YML063W          -4.367071   
113        4214           RGI1     YER067W           5.637591   
114        4215           RGI2     YIL057C           4.394700   
115        4586         RPL24A     YGL031C          -4.649040   
116        4587         RPL24B     YGR148C          -3.636010   
117        4764           GLG2     YJL137C           2.054324   
118        4765           GLG1     YKR058W           2.885785   
119        4814        YHL012W     YHL012W           3.564791   
120        4815           UGP1     YKL035W           2.983799   
121        4826           PYC2     YBR218C          -0.280954   
122        4827           PYC1     YGL062W           0.842286   
123        4870           IMD4     YML056C          -4.749147   
124        4871           IMD3     YLR432W          -3.783399   
125        4900         RPL16A     YIL133C          -4.138627   
126        4901         RPL16B     YNL069C          -4.370756   
127        4936         RPL22B   YFL034C-A          -3.041141   
128        4937         RPL22A     YLR06

In [33]:
goi_data_phylogenies.to_csv(data_processing_dir + os.sep + "20170817_klscpka" + os.sep + "goi_data_phylogenetic_trees.csv")   
    

In [9]:
phylome_db_name = sc_phylomedb_convert[sc_phylomedb_convert['sc_genename']==gene]['sc_genename'].values[0]

In [11]:
phylome_db_name.values[0]

'YKL109W'

In [71]:
#Add topology columns to dataframe

ortholog_dataset['joint_topologies'] = joint_topologies

for fc_level in ['low','high']:
    topologies_fc_level = []
    for topology in topologies: 
        topology_fc_level = {}
        for fc_level_tree in ['low','high']:
            if isinstance(topology[fc_level_tree],str):
                topology_fc_level[fc_level_tree] = None
            elif isinstance(topology[fc_level_tree],dict):
                topology_fc_level[fc_level_tree] = topology[fc_level_tree][fc_level]
        topologies_fc_level.append(topology_fc_level)
    ortholog_dataset['topologies_'+ fc_level] = topologies_fc_level


In [14]:
ortholog_dataset.columns

Index(['Number', 'Ancestor', 'AA %id', 'Length Ratio',
       'SC_common_name_YGOB_low', 'N SCer Orthologs_low', 'kl_genename_low',
       'sc_genename_low', 'baseMean_KL_low', 'log2FoldChange_KL_low',
       'lfcSE_KL_low', 'stat_KL_low', 'pvalue_KL_low', 'padj_KL_low',
       'rlog_AS_noNMPP1_YPD_t40_mean_KL_low',
       'rlog_AS_yesNMPP1_YPD_t40_mean_KL_low',
       'rlog_AS_noNMPP1_YPD_t40_var_KL_low',
       'rlog_AS_yesNMPP1_YPD_t40_var_KL_low',
       'raw_AS_noNMPP1_YPD_t40_mean_KL_low',
       'raw_AS_yesNMPP1_YPD_t40_mean_KL_low',
       'raw_AS_noNMPP1_YPD_t40_var_KL_low',
       'raw_AS_yesNMPP1_YPD_t40_var_KL_low', 'baseMean_SC_low',
       'log2FoldChange_SC_low', 'lfcSE_SC_low', 'stat_SC_low', 'pvalue_SC_low',
       'padj_SC_low', 'rlog_AS_noNMPP1_YPD_t40_mean_SC_low',
       'rlog_AS_yesNMPP1_YPD_t40_mean_SC_low',
       'rlog_AS_noNMPP1_YPD_t40_var_SC_low',
       'rlog_AS_yesNMPP1_YPD_t40_var_SC_low',
       'raw_AS_noNMPP1_YPD_t40_mean_SC_low',
       'raw_AS_yesNMP

In [72]:
#How many get consistent results from both low and high trees? 

topologies_gene_conv = {'A-A1','B-B1','C-C1','C-C2'}
topologies_diff_lineages = {'A-A2','B-B2','A-C','A-B','B-C'}

resolution = 'exact'  #options are 'exact' or 'type'

ind_list = []
for ind,row in ortholog_dataset.iterrows(): 
    joint_topology = row['joint_topologies']
    #Make sure none of the joint topologies are pathological or None
    if len({None, 'pathological'} - {joint_topology['high'],joint_topology['low']})==2:
        #check to see if both joint topologies agree
        if resolution == 'exact':
            if joint_topology['low']==joint_topology['high']:
                ind_list.append(ind)
                print("{} and {} have matching joint topologies: {} and {} respectively".format(row['SC_common_name_low'],row['SC_common_name_high'],joint_topology['low'],joint_topology['high']))           
        elif resolution == 'type': 
            joint_topology_type = {'low':io_library.joint_topology_type(joint_topology['low']), 
                                   'high': io_library.joint_topology_type(joint_topology['high'])}
            if joint_topology_type['low']==joint_topology_type['high']:
                ind_list.append(ind)
                print("{} and {} have similar joint topologies: {} and {} respectively".format(row['SC_common_name_low'],row['SC_common_name_high'],joint_topology['low'],joint_topology['high']))


            

columns_to_view = ['SC_common_name_low','SC_common_name_high','log2FoldChange_SC_low', 'log2FoldChange_SC_high', 'log2FoldChange_KL_low','joint_topologies','topologies_low','topologies_high','threshold_group_KL','threshold_group_SC']
ortholog_dataset.loc[ind_list,columns_to_view]


#exactly equal 
#   How many are "gene_conversion" 
#   How many are "different_lineages"

SPS2 and SPS22 have matching joint topologies: C-outgroup-? and C-outgroup-? respectively
EUG1 and PDI1 have matching joint topologies: C-outgroup-? and C-outgroup-? respectively
GNP1 and AGP1 have matching joint topologies: A-A1 and A-A1 respectively
TOS4 and PLM2 have matching joint topologies: A-C and A-C respectively
RPL37A and RPL37B have matching joint topologies: A-A1 and A-A1 respectively
SKG3 and CAF120 have matching joint topologies: B-C and B-C respectively
LAM1 and SIP3 have matching joint topologies: A-A2 and A-A2 respectively
PHO90 and PHO87 have matching joint topologies: A-A1 and A-A1 respectively
RPL17B and RPL17A have matching joint topologies: A-A1 and A-A1 respectively
PIR1 and PIR5 have matching joint topologies: C-outgroup-? and C-outgroup-? respectively
RPS21B and RPS21A have matching joint topologies: A-A1 and A-A1 respectively
CAF4 and MDV1 have matching joint topologies: C-outgroup-? and C-outgroup-? respectively
SET3 and SET4 have matching joint topologies: C

SC_common_name_low SC_common_name_high  log2FoldChange_SC_low  \
3                 SPS2               SPS22              -1.076672   
4                 EUG1                PDI1              -0.030605   
9                 GNP1                AGP1              -2.082446   
13                TOS4                PLM2              -1.835652   
14              RPL37A              RPL37B              -4.388579   
15                SKG3              CAF120              -0.207497   
17                LAM1                SIP3               0.440538   
18               PHO90               PHO87              -2.928895   
22              RPL17B              RPL17A              -4.669997   
25                PIR1                PIR5              -0.161061   
30              RPS21B              RPS21A              -4.060328   
34                CAF4                MDV1              -0.814438   
36                SET3                SET4               0.232533   
37                BCH2                CHS6              -0.324363   
39                ALY1                ALY2               0.610060   
41                IML2             YKR018C               1.181692   
46                UBX6                UBX7               0.666153   
52                ESC8                IOC3              -1.878378   
53                CMK2                CMK1               0.486755   
54                GSY2                GSY1               4.175972   
55                RCK2                RCK1               1.943103   
56                YEF3                HEF3              -5.379224   
57                IDP3                IDP2               1.281757   
58                HRB1                GBP2              -0.885933   
62                NPP2                NPP1              -0.581341   
63                GEA1                GEA2              -1.474240   
64               ECM10                SSC1               0.616683   
65                HYP2                ANB1              -1.241898   
66                CYC1                CYC7              -1.725706   
67                UTR1                YEF1               0.411152   
..                 ...                 ...                    ...   
464            YDR132C             YLR108C              -1.464658   
470               SDH4                SHH4               1.447171   
471               UPS2                UPS3              -0.181717   
472              FAR10               VPS64               0.363791   
473               EBS1                EST1              -0.914213   
474              ECM22                UPC2              -0.725425   
475            YDR222W             YLR225C              -3.248088   
478               SKS1                VHS1              -0.713168   
480               EGD1                BTT1              -1.302627   
483               KTR6                KRE2              -0.599051   
484               SUR1                CSH1               0.054169   
486             RPL21B              RPL21A              -3.801607   
488               YDC1                YPC1               1.778761   
491               EEB1                EHT1              -0.659105   
493               SSE1                SSE2              -1.083654   
495               DED1                DBP1              -1.140894   
496               ODC2                ODC1              -2.194926   
497               ISU2                ISU1              -1.979223   
498               GIP3                HER1              -1.008489   
499               WTM2                UME1              -0.208082   
500               MKK1                MKK2              -0.230907   
502             RPL33A              RPL33B              -3.619488   
503               KES1                HES1              -0.332369   
507              UAF30                TRI1              -1.131872   
508               BUD7                BCH1              -1.547439   
509               ZRC1                COT1              -0.809731   
51

Looking for exact matches of joint topologies, there are 221 of my entire ortholog set using TORDC and KLULA as seed species that match.  Marcet-Houben and Galbadon had 463 that they kept.

['A-A1 : 83 37.556561% ',  (33.91%) 
 'A-C : 40 18.099548% ', (8.64%)
 'B-C : 13 5.882353% ',  (0.86%) 
 'A-A2 : 38 17.194570% ', (43.63%)
 'C-C2 : 35 15.837104% ', (3.46%)
 'B-B1 : 7 3.167421% ',  (3.89%))
 'C-C1 : 5 2.262443% '] (4.75%)
 B-B2: 0 0%  (0.86%)
 
 I seem to be underestimating A-A2, and overestimating C-C2, B-C, and A-C. 


In [73]:
ortholog_subset = ortholog_dataset.loc[ind_list,:]
joint_topology_summary = Counter([joint_topology['low'] for joint_topology in ortholog_subset['joint_topologies']])
total = len(ortholog_subset)
["{} : {} {:0%} ".format(key, val, val/total) for key, val in joint_topology_summary.items()]



['C-outgroup-? : 57 23.170732% ',
 'A-A1 : 83 33.739837% ',
 'A-C : 25 10.162602% ',
 'B-C : 9 3.658537% ',
 'A-A2 : 37 15.040650% ',
 'C-C2 : 23 9.349593% ',
 'B-B1 : 7 2.845528% ',
 'C-C1 : 5 2.032520% ']

In [44]:
#Visualize with a heatmap

ortholog_dataset_subset = ortholog_dataset[ortholog_dataset['kl_genename_low'].isin(ohnologs_goi['kl_genename_low'])] #ortholog_dataset 

ortholog_dataset_subset = ortholog_dataset_subset.sort_values(by = ['threshold_group_SC','log2FoldChange_KL_low'])

#fold change values
fc_values = ortholog_dataset_subset.loc[:,['log2FoldChange_SC_low', 'log2FoldChange_SC_high', 'log2FoldChange_KL_low']]


#Topology values
WGH_branch_seed = 'YEAST'
KLE_branch_seed = 'KLULA'
ZT_branch_seed = 'TORDC'


joint_topology_values = []
joint_topology_column_labels = []
joint_topology_text = []

#Visualize joint column low and high
value_dict_joint = {'A-B':7,
              'A-A2':8,
              'B-B2':9,
              'A-C': 10,
              'B-C': 11,
              None: -7,
              'pathological': -6,
              'B-B1':0,
              'A-A1':1,
              'C-C1':2,
              'C-C2':3,
              }


for fc_level in ['low','high']:
    joint_topologies_text_fc_level = [joint_topology[fc_level] for ind, joint_topology in ortholog_dataset_subset['joint_topologies'].items()]
    
    joint_topology_text.append(joint_topologies_text_fc_level)
    joint_topology_values.append([value_dict_joint[item] for item in joint_topologies_text_fc_level])
    
    joint_topology_column_labels.append('joint topology ' + fc_level)

value_dict = {'A':13, 
              'A_late': 12, 
              'B_early': 11,
              'B':0,
              'B_late': 1,
              'A_early': 2,
              'C':7,
              'C_early':6,
              'C_late':5,
              'under construction':-5,
              'missing ' + KLE_branch_seed + ' protein':-6.5,
              'missing ' + ZT_branch_seed + ' and ' + KLE_branch_seed +' protein':-6.5,
              'missing ' + ZT_branch_seed + ' protein':-6.5,
               None:-7
                     }

topology_values = []
topology_column_labels = []
topology_text = []
for fc_level in ['low','high']:
    for fc_level_tree in ['low','high']:
        topology_values.append([value_dict[topology[fc_level_tree]] for ind,topology in ortholog_dataset_subset['topologies_'+fc_level].items()])
        topology_text.append([topology[fc_level_tree] for ind,topology in ortholog_dataset_subset['topologies_'+fc_level].items()])
        topology_column_labels.append('topology ' + fc_level + ' (from ' + fc_level_tree + ' tree)')


#Visualize low and high low and high for each base tree



# for fc_group in ['low', 'high']:
#     for topology_column in topology_columns: 
#         print(fc_group + " " + topology_column)
#         KLE_branch_seed = topology_column.split('_')[2]
#         ZT_branch_seed = topology_column.split('_')[3]
#         value_dict = {'A':-5, 'B':5,'C':0,
#                       'under construction':3,
#                       'missing ' + KLE_branch_seed + ' protein':-1,
#                       'missing ' + ZT_branch_seed + ' and ' + KLE_branch_seed +' protein':-2,
#                       'missing ' + ZT_branch_seed + ' protein':-1.5,
#                        np.nan:-2.5
#                      }
#         topology_values.append([value_dict[topology] for topology in ohnologs_topology_sorted[topology_column + '_' + fc_group]])
        
#         topology_column_labels = WGH_branch_seed[0] + KLE_branch_seed[0] + ZT_branch_seed[0] + '_' + fc_group


gene_labels = ortholog_dataset_subset['SC_common_name_low']+ ' / ' + ortholog_dataset_subset['SC_common_name_high']

trace1 = pygo.Heatmap(z = np.array(fc_values),
                   x=['SC low FC', 'SC high FC', 'KL FC'],
                   y= gene_labels
                   )

trace2 = pygo.Heatmap(z = np.array(joint_topology_values).transpose(),
                   text = np.array(joint_topology_text).transpose(),
                   x=joint_topology_column_labels,
                   y= gene_labels, 
                   hoverinfo = 'x+y+text',
                   colorscale = 'Electric',
                   showscale = False)

trace3 = pygo.Heatmap(z = np.array(topology_values).transpose(),
                      text = np.array(topology_text).transpose(),
                   x=topology_column_labels,
                   y= gene_labels,
                   colorscale = 'Blackbody',
                   hoverinfo = 'x+y+text',
                   showscale = False)


fig = pytools.make_subplots(rows=1, cols=3)

fig.append_trace(trace1, 1, 1)
fig.append_trace(trace2, 1, 2)
fig.append_trace(trace3, 1, 3)

#fig['layout'].update(height=600, width=600, title='i <3 subplots')

# layout = pygo.Layout(
#     yaxis = {'title': 'Low Paralog / High Paralog'}
# )

#fig = pygo.Figure(data=data, layout=layout)
py.iplot(fig, filename='topology_heatmap')

This is the format of your plot grid:
[ (1,1) x1,y1 ]  [ (1,2) x2,y2 ]  [ (1,3) x3,y3 ]



In [ ]:
#Separate out data that has a clear interpretation

In [25]:
fc_level = 'low'
Counter([topology[fc_level] for ind,topology in ortholog_dataset['topologies_low'].items()])

Counter({'missing KLULA protein': 18,
         'C': 50,
         None: 112,
         'missing TORDC protein': 34,
         'A': 139,
         'under construction': 15,
         'C_early': 29,
         'B': 27,
         'A_early': 31,
         'C_late': 31,
         'B_late': 2,
         'A_late': 28,
         'B_early': 3})

In [79]:
fc_level_tree = 'low'
fc_level
[value_dict[topology[fc_level_tree]] for ind,topology in a['topologies_'+fc_value].items()]

[5]

In [78]:
columns_to_view = ['joint_topologies','topologies_low','topologies_high']
a = ortholog_dataset_subset[ortholog_dataset_subset['SC_common_name_low']=='EMI2'].loc[:,columns_to_view]

In [45]:
joint_topologies_equal = []
for ind, row in ortholog_dataset.iterrows():
    joint_topologies_equal.append((row['joint_topologies']['low']) == (row['joint_topologies']['high']))

ortholog_dataset_subset = ortholog_dataset[joint_topologies_equal]


In [73]:
ortholog_dataset_subset = ortholog_dataset[ortholog_dataset['kl_genename_low'].isin(ohnologs_goi['kl_genename_low'])]

In [ ]:
[{'high': None, 'low': 'missing KLULA protein'},
 {'high': 'A', 'low': 'C'},
 {'high': None, 'low': None},
 {'high': 'missing TORDC protein', 'low': 'missing TORDC protein'},
 {'high': 'C', 'low': 'C'},

In [123]:
phylomedb_name_list = ['Phy000CX94_YEAST', 'Phy000CZ5A_YEAST', 'Phy001SPOB_YEAST', 'Phy000CWKV_YEAST', 'Phy000CWKW_YEAST', 'Phy000CY0K_YEAST', 'Phy000CZRB_YEAST','Phy000CXWG_YEAST', 'Phy000CWKY_YEAST','Phy000CXWI_YEAST',  'Phy000CY0P_YEAST', 'Phy000CYP9_YEAST']
io_library.SC_common_name_lookup(sc_phylomedb_convert[sc_phylomedb_convert['phylomedb_name'].isin(phylomedb_name_list)]['sc_genename'])


['HXT11',
 'HXT7',
 'HXT6',
 'HXT4',
 'HXT8',
 'HXT9',
 'HXT3',
 'GAL2',
 'HXT2',
 'HXT5',
 'HXT1',
 'HXT10']

In [120]:
tree = Tree(trees.loc['Phy000CWKY_YEAST','tree'])
print(tree)


                                             /-Phy004F6AS_1071379
                                            |
                                          /-|      /-Phy004F6Z9_1071379
                                         |  |   /-|
                                         |   \-|   \-Phy004F6V0_1071379
                                       /-|     |
                                      |  |      \-Phy004F74K_1071379
                                      |  |
                                      |   \-Phy004F9A6_1071379
                                      |
                                      |                  /-Phy004F6C5_1071379
                                      |               /-|
                                      |            /-|   \-Phy004F9IB_1071379
                                    /-|           |  |
                                   |  |         /-|   \-Phy004F77S_1071379
                                   |  |        |  |
                               

In [181]:
common_name = 'MPC2'
tree = Tree(trees_subset[common_name])
gene = phylomedb_names[phylomedb_names['SC_common_name']==common_name]['phylomedb_name'].iloc[0]
# print(tree)

#Extract only leaves that are are _YEAST, _KLULA, ZT_branch seed (i.e. _TORDC, _ZYGRO)
leaf_names = tree.get_leaf_names()
leaf_subset = [name for name in leaf_names if name.split('_')[1] in {"YEAST","KLULA",ZT_branch_seed}]

#Check for 1-2 S.Cer proteins and only 1 each of KL and ST branch
leaf_subset_spec = [name.split("_")[1] for name in leaf_subset]
leaf_subset_spec_counter = Counter(leaf_subset_spec)

leaf_subset_spec_counter

for WGH_branch in ['KLULA',ZT_branch_seed]:
    if (leaf_subset_spec_counter[WGH_branch]>1):
        #Throw out extra WGH branch gene if children of common ancestor between it and current gene are in an outgroup.  
        
        #Prune down tree to include spec of interest and outgroups
        leaf_subset_outgroups = [name for name in leaf_names if name.split('_')[1] in set(["YEAST","KLULA",ZT_branch_seed] + outgroups)]
        tree.prune(leaf_subset_outgroups) 
        #print(tree)
        #for each of the branch genes. 
        branch_genes = [leaf for ind,leaf in enumerate(leaf_subset) if leaf_subset_spec[ind]==WGH_branch]
        for branch_gene in branch_genes:
            ancestor = tree.get_common_ancestor([gene,branch_gene])
            ancestor_leaves = ancestor.get_leaf_names()
            ancestor_leaves_spec = [name.split("_")[1] for name in ancestor_leaves]
    
            #if the leaves of the parent node include an outgroup, throw out the gene
            print(ancestor_leaves_spec)
            if len(set(ancestor_leaves_spec) & set(outgroups)) > 0:
                print('Threw out ' + branch_gene + ' because outgroup gene between it and seed')
                leaf_subset.remove(branch_gene)    
                leaf_subset_spec = [name.split("_")[1] for name in leaf_subset]
                leaf_subset_spec_counter = Counter(leaf_subset_spec)
                
            if  (leaf_subset_spec_counter[WGH_branch]==0):
                print('Alert threw out both ' + WGH_branch + ' genes for seed: ' + gene)
    
    
# # #Extract those ones plus outgroups
# # leaf_names = tree.get_leaf_names()
# # 
# # print(tree)
# gene = ZT_genes[0]



#Warn if it throws out both Zygro genes


['YEAST', 'ZYGRO', 'YEAST', 'KLULA']
['DEKBR', 'YEAST', 'KLULA', 'DEKBR', 'YEAST', 'ZYGRO', 'YEAST', 'KLULA', 'CANAL', 'YARLI']
Threw out Phy0008N7U_KLULA because outgroup gene between it and seed


In [13]:
joint_topologies

[{'high': None, 'low': 'pathological'},
 {'high': 'A-A1', 'low': 'C-C1'},
 {'high': None, 'low': None},
 {'high': 'pathological', 'low': 'pathological'},
 {'high': 'A-C', 'low': 'B-C'},
 {'high': 'A-A2', 'low': 'A-A1'},
 {'high': None, 'low': None},
 {'high': 'pathological', 'low': 'A-C'},
 {'high': 'pathological', 'low': 'pathological'},
 {'high': 'A-A1', 'low': 'A-A1'},
 {'high': 'C-C2', 'low': 'A-A1'},
 {'high': 'B-C', 'low': 'C-C2'},
 {'high': 'C-C2', 'low': 'B-C'},
 {'high': 'A-C', 'low': 'A-C'},
 {'high': 'A-A1', 'low': 'A-A1'},
 {'high': 'B-C', 'low': 'B-C'},
 {'high': 'pathological', 'low': 'pathological'},
 {'high': 'A-A2', 'low': 'A-A2'},
 {'high': 'A-A1', 'low': 'A-A1'},
 {'high': 'B-C', 'low': 'C-C2'},
 {'high': 'A-A2', 'low': 'B-B2'},
 {'high': 'A-C', 'low': None},
 {'high': 'A-A1', 'low': 'A-A1'},
 {'high': 'pathological', 'low': 'A-C'},
 {'high': 'A-C', 'low': 'A-B'},
 {'high': 'pathological', 'low': 'pathological'},
 {'high': 'pathological', 'low': 'pathological'},
 {'h

In [30]:
#Add topology columns to ortholog database
topology_columns = [column for column in trees_topology.columns if column.split('_')[0]=='topologies']
#sc_phylomedb_convert_topology = pd.merge(sc_phylomedb_convert, pd.DataFrame(trees[topology_columns]), how = 'left', left_on = 'phylomedb_name', right_index = True)
ohnologs_topology_gene1 = pd.merge(ohnologs_goi, trees_topology, how = 'left', left_on = 'sc_genename_low', right_on = 'sc_genename' )
ohnologs_topology = pd.merge(ohnologs_topology_gene1, trees_topology, how = 'left', left_on = 'sc_genename_high', right_on = 'sc_genename',suffixes = ['_low', '_high'])


In [38]:
#Finds all possible values for topology for a given set of columns. 
list(set(list(ohnologs_topology[topology_columns[1] + '_low'])+list(ohnologs_topology[topology_columns[1] + '_high'])))


[nan,
 'A',
 'B',
 'under construction',
 'missing KLULA protein',
 'missing ZYGRO and KLULA protein',
 'C']

In [200]:
#Plot out summaries for all of these
print(Counter(ohnologs_topology_sorted['topologies_low']))
print(Counter(ohnologs_topology_sorted['topologies_high']))

Counter({'A': 31, 'C': 14, 'B': 4, 'missing KLULA protein': 4, 'missing TORDC and KLULA protein': 4, 'under construction': 3, nan: 1})
Counter({'A': 26, 'C': 22, 'missing KLULA protein': 4, 'B': 2, 'under construction': 2, 'missing TORDC and KLULA protein': 2, nan: 2, 'missing TORDC protein': 1})


In [187]:
print(Counter(ohnologs_topology_sorted['topologies_low']))
print(Counter(ohnologs_topology_sorted['topologies_high']))

Counter({'A': 32, 'C': 14, 'B': 4, 'missing KLULA protein': 4, 'missing ZYGRO and KLULA protein': 4, 'under construction': 2, nan: 1})
Counter({'A': 25, 'C': 23, 'missing KLULA protein': 4, 'under construction': 3, 'B': 2, 'missing ZYGRO and KLULA protein': 2, nan: 2})


In [85]:
#look at tree topology for given sc common name
gois = ['GDH1','GDH3','CDC19','PYK2','MPC2','MPC3']

phylomedb_names = sc_phylomedb_convert[sc_phylomedb_convert['sc_genename'].isin(io_library.SC_orf_lookup_by_name(gois))].copy()

phylomedb_names['SC_common_name'] = io_library.SC_common_name_lookup(phylomedb_names['sc_genename'])

trees_subset = {}
for ind, row in phylomedb_names.iterrows():
    trees_subset[row['SC_common_name']]=trees.loc[row['phylomedb_name'],'tree']






In [1]:
columns_to_view = ['Gene 1', 'Gene Name 1', 'log2FoldChange_SC_gene1','Gene 2', 'Gene Name 2',  'log2FoldChange_SC_gene2', 'threshold_group_SC','kl_genename_gene1','log2FoldChange_KL_gene1','threshold_group_KL']
ohnologs_expression[ohnologs_expression['Gene Name 1'].isin(['SDC25','YDR134C','YGR273C'])].loc[:,columns_to_view]


NameError: name 'ohnologs_expression' is not defined

In [37]:
phylomedb_names = ['Phy000CZF5_YEAST', 'Phy000CX7L_YEAST', 'Phy000CZ7U_YEAST']
sc_phylomedb_convert_topology[sc_phylomedb_convert_topology['phylomedb_name'].isin(phylomedb_names)]

phylomedb_name sc_genename          topologies
288   Phy000CZ7U_YEAST     YMR096W  under construction
3578  Phy000CZF5_YEAST     YNL333W  under construction
3775  Phy000CX7L_YEAST     YFL059W  under construction

In [12]:
ohnologs_topology[ohnologs_topology['Gene Name low'].isin(['GDH1','GDH3'])]
#['Gene 1', 'Gene Name 1', 'log2FoldChange_SC_gene1','Gene 2', 'Gene Name 2',  'log2FoldChange_SC_gene2', 'threshold_group_SC','kl_genename_gene1','log2FoldChange_KL_gene1','threshold_group_KL']

threshold_group_SC threshold_group_KL kl_genename_gene1  \
379          down_flat               flat      KLLA0F00594g   

     log2FoldChange_KL_gene1 Gene low Gene Name low  log2FoldChange_SC_low  \
379                 -0.50328  YOR375C          GDH1              -3.521635   

    Gene high Gene Name high  log2FoldChange_SC_high phylomedb_name_low  \
379   YAL062W           GDH3                 2.46504   Phy000D070_YEAST   

    sc_genename_low topologies_low phylomedb_name_high sc_genename_high  \
379         YOR375C              A    Phy000CVIN_YEAST          YAL062W   

    topologies_high  
379               A

In [17]:
#filter to up_flat in SC and flat in KL.
ohnologs_topology_subset = ohnologs_topology[(ohnologs_topology['threshold_group_SC']=='up_flat') & (ohnologs_topology['threshold_group_KL']=='flat')]

ohnologs_topology_subset[ohnologs_topology_subset['Gene Name low'].isin(['CDC19','PYK2','MPC2','MPC3'])]

threshold_group_SC threshold_group_KL kl_genename_gene1  \
261            up_flat               flat      KLLA0A05500g   
383            up_flat               flat      KLLA0F23397g   

     log2FoldChange_KL_gene1 Gene low Gene Name low  log2FoldChange_SC_low  \
261                -1.789380  YHR162W          MPC2              -1.860001   
383                 0.086733  YAL038W         CDC19              -0.908989   

    Gene high Gene Name high  log2FoldChange_SC_high phylomedb_name_low  \
261   YGR243W          FMP43                3.387496   Phy000CXYH_YEAST   
383   YOR347C           PYK2                2.544465   Phy000CVJD_YEAST   

    sc_genename_low      topologies_low phylomedb_name_high sc_genename_high  \
261         YHR162W  under construction    Phy000CXQ1_YEAST          YGR243W   
383         YAL038W                   C    Phy000D068_YEAST          YOR347C   

        topologies_high  
261  under construction  
383                   C

In [50]:
ohnologs[ohnologs['Gene 2']=='YCR075W-A']

Number   Ancestor     Gene 1 Gene Name 1     Gene 2 Gene Name 2 AA %id  \
396     397  Anc_6.348  YNR034W-A   YNR034W-A  YCR075W-A   YCR075W-A    18%   

     Length Ratio  
396          0.77

In [51]:
ohnologs_topology[ohnologs_topology['Gene Name low']=='YCR075W-A']

threshold_group_SC threshold_group_KL kl_genename_gene1  \
376            up_flat               flat      KLLA0F27390g   

     log2FoldChange_KL_gene1   Gene low Gene Name low  log2FoldChange_SC_low  \
376                 0.257658  YCR075W-A     YCR075W-A               0.542983   

     Gene high Gene Name high  log2FoldChange_SC_high phylomedb_name_low  \
376  YNR034W-A      YNR034W-A                7.183301   Phy000CW2F_YEAST   

    sc_genename_low topologies_low phylomedb_name_high sc_genename_high  \
376       YCR075W-A              C    Phy000CZPR_YEAST        YNR034W-A   

        topologies_high  
376  under construction

In [33]:
Counter(ohnologs_topology_subset['topologies_low'])

Counter({'under construction': 40, nan: 3, 'C': 13, 'A': 22, 'B': 2})

In [34]:
Counter(ohnologs_topology_subset['topologies_high'])

Counter({'under construction': 40, 'C': 17, 'A': 17, 'B': 1, nan: 5})

In [192]:
hover_text = []
for ind, row in ohnologs_topology_subset.iterrows():
    #(row['Gene Name 1'],row['topologies_gene1'],row['Gene Name 2'], row['topologies_gene2'])
    hover_text.append('{}<br>{}<br>{}<br>{}'.format(row['Gene Name 1'],row['topologies_gene1'],row['Gene Name 2'], row['topologies_gene2']))

#titles = {1: 'One S.Cer ortholog ', 2: 'Two S.Cer orthologs '}
#colors = {1: 'grey', 2: 'red'}

trace = pygo.Scatter(
            x= ohnologs_topology_subset['log2FoldChange_SC_gene1'], #[1,2,3,4],
            y= ohnologs_topology_subset['log2FoldChange_SC_gene2'], #[1.1,2,3.5,7],
            mode= 'markers',
            text= hover_text #['a','b','c','d']
            )

layout = pygo.Layout(
    title= 'topology for ohnologs',
    hovermode= 'closest',
    xaxis= {'title':'gene 1 fold change'},
    yaxis= {'title':'gene 2 fold change'}
    )
    
fig= pygo.Figure(data=[trace], layout=layout)


# N = 1000
# random_x = np.random.randn(N)
# random_y = np.random.randn(N)

# # Create a trace
# trace = pygo.Scatter(
#     x = random_x,
#     y = random_y,
#     mode = 'markers',
#     text = random_x
# )

# data = [trace]

# layout = pygo.Layout(
#     title= 'Stats of USA States',
#     hovermode= 'closest'
#     )

# fig= pygo.Figure(data=data, layout=layout)

# Plot and embed in ipython notebook!
#py.iplot(data, filename='basic-scatter')

#py.iplot(fig, filename='SC_foldchange_topology_overlay')
py.iplot(fig, filename='SC_foldchange_topology_overlay')


# fig_data = {
#   			'x': [1,2,3,4.5],    #list(ohnologs_topology_subset['log2FoldChange_SC_gene1']), 
#         	'y': [2,3,4,5.7],    #list(ohnologs_topology_subset['log2FoldChange_SC_gene2']) 
#             'mode': 'markers'
#             } #, 
#         	'text': hover_text, 
#         	'mode': 'markers', 
#             'marker': {
#                 'opacity': 0.5,
#                 #'color': colors[NN]
#                 }
#             }
        	#'name': } 

# fig = {
#     'data': fig_data
#     'layout': {
#                 'xaxis': {'title': 'Gene 1 fold change'},
#                 'yaxis': {'title': 'Gene 2 fold change'}
#                }
       





In [152]:
#filter to genes with separated expression
foldchange_separation = 3.5
ohnologs_topology_separated = ohnologs_topology[abs(ohnologs_topology['log2FoldChange_SC_gene1'] - ohnologs_topology['log2FoldChange_SC_gene2']) > foldchange_separation]


columns_to_view = ['Gene 1', 'Gene Name 1', 'log2FoldChange_SC_gene1','topologies_gene1', 'Gene 2', 'Gene Name 2',  'log2FoldChange_SC_gene2', 'topologies_gene2','threshold_group_SC','log2FoldChange_KL_gene1','threshold_group_KL']
#['Gene 1', 'Gene Name 1', 'kl_genename_gene1', 'Gene 2', 'Gene Name 2', 'kl_genename_gene2', 'log2FoldChange_SC_gene1', 'log2FoldChange_SC_gene2', 'threshold_group_SC','log2FoldChange_KL_gene1','threshold_group_KL', 'topologies_gene1', 'topologies_gene2']
ohnologs_topology_separated.loc[:,columns_to_view]


Gene 1 Gene Name 1  log2FoldChange_SC_gene1    topologies_gene1  \
0      YCL051W        LRE1                 1.006787                   C   
2    YDR524C-B   YDR524C-B                -2.236981                 NaN   
6      YCL037C        SRO9                -4.754452  under construction   
19     YCR034W        FEN1                -2.858003  under construction   
24     YJL164C        TPK1                 2.844732  under construction   
26     YJL159W      HSP150                 0.711945  under construction   
35     YJL110C        GZF3                -1.197971  under construction   
42     YJL079C        PRY1                 1.883493                 NaN   
50     YFR017C        IGD1                 5.532256  under construction   
56     YLR249W        YEF3                -5.379224  under construction   
60     YCR005C        CIT2                -2.464709  under construction   
66     YEL039C        CYC7                 5.703798                   C   
69     YJR054W     YJR054W                -1.852029                   C   
92     YNL104C        LEU4                -0.167463  under construction   
96     YNL093W       YPT53                 5.742324  under construction   
97     YNL087W        TCB2                -3.271993  under construction   
106    YIL113W        SDP1                 5.686275                   B   
114    YDL085W        NDE2                 7.299282  under construction   
119    YLR029C      RPL15A                -4.289172                   A   
122    YKL141W        SDH3                 1.507281                   B   
126    YKL127W        PGM1                -0.703705                   A   
132    YKL091C     YKL091C                 4.711886  under construction   
146    YKL072W        STB6                -1.920920                   A   
149    YNL309W        STB1                -0.688053                   C   
160    YBR147W        RTC2                 5.373333  under construction   
167    YDL022W        GPD1                 3.828820                   B   
168    YDL021W        GPM2                 5.378055  under construction   
176    YBR042C       CST26                -0.729865                   C   
181  YDR034W-B   YDR034W-B                 6.620033                 NaN   
188    YBR104W        YMC2                -4.971720  under construction   
..         ...         ...                      ...                 ...   
331    YLR273C        PIG1                 1.755991  under construction   
334    YLR293C        GSP1                -3.170040  under construction   
337    YBR214W       SDS24                 4.263882                   C   
357    YMR177W        MMT1                -2.651547                   C   
360    YMR182C        RGM1                 0.506873  under construction   
376  YNR034W-A   YNR034W-A                 7.183301  under construction   
379    YAL062W        GDH3                 2.465040                   A   
385    YAL034C       FUN19                 2.563213  under construction   
398    YER054C        GIP2                 5.784015  under construction   
402    YER062C        HOR2                 1.811356  under construction   
405    YER070W        RNR1                -3.323715  under construction   
408    YIL079C        AIR1                -4.834911                   B   
419    YBL054W        TOD6                -4.042864                   C   
422    YBL061C        SKT5                -1.806729                   A   
424    YBL068W        PRS4                -3.370891  under construction   
437    YBL039C        URA7                -6.543049                   A   
438    YBL043W       ECM13                -1.705201                 NaN   
442    YFL048C       EMP47                 0.038517                   A   
454    YDR077W        SED1                 1.686499                 NaN   
460    YDR107C        TMN2                 1.495041  under construction   
470    YDR178W        SDH4                 1.447171  under construction   
475    YDR222W     YDR222W       

In [9]:
#Convert ohnologs_expression to a sorted version. 

#loop through each item. 

columns_by_gene_no = {'gene1': ['Gene 1', 'Gene Name 1', 'log2FoldChange_SC_gene1'], 'gene2': ['Gene 2', 'Gene Name 2',  'log2FoldChange_SC_gene2']}

#make new columns: 
new_column_list = ['Gene low', 'Gene Name low', 'log2FoldChange_SC_low', 'Gene high', 'Gene Name high', 'log2FoldChange_SC_high']

new_columns_data = []

for index, row in ohnologs_expression.iterrows():
    #Decide if gene1 or gene2 is low expression
    if row['log2FoldChange_SC_gene1']<row['log2FoldChange_SC_gene2']:
        low_gene = 'gene1'
        high_gene = 'gene2'
    elif  row['log2FoldChange_SC_gene1']>row['log2FoldChange_SC_gene2']:
        low_gene = 'gene2'
        high_gene = 'gene1'
    else:
        print('problems with {} and {}'.format(row['Gene Name 1'],row['Gene Name 2']))
    
    new_columns_data.append([row[column] for column in columns_by_gene_no[low_gene]] + [row[column] for column in columns_by_gene_no[high_gene]])

#existing columns to keep
columns_to_keep = ['threshold_group_SC','threshold_group_KL', 'kl_genename_gene1','log2FoldChange_KL_gene1']
ohnologs_expression_sorted = ohnologs_expression.loc[:,columns_to_keep]

#this seems to be looping through inefficiently
for ind,new_column_name in enumerate(new_column_list):
    ohnologs_expression_sorted[new_column_name] = [row[ind] for row in new_columns_data]

ohnologs_expression_sorted

# #['Gene 1', 'Gene Name 1', 'log2FoldChange_SC_gene1','Gene 2', 'Gene Name 2',  'log2FoldChange_SC_gene2', 'threshold_group_SC','kl_genename_gene1','log2FoldChange_KL_gene1','threshold_group_KL']
# ohnologs_expression.columns

problems with SDC25 and CDC25
problems with YGR273C and YMR295C
problems with YDR134C and CCW12


threshold_group_SC threshold_group_KL kl_genename_gene1  \
0            down_flat               flat      KLLA0C00803g   
1            flat_flat               flat      KLLA0C00847g   
2              up_flat               flat      KLLA0C00957g   
3            flat_flat                 up      KLLA0C01001g   
4            flat_flat               flat      KLLA0C01111g   
5              up_flat               flat      KLLA0C01155g   
6            down_flat               down      KLLA0C01210g   
7            down_flat               flat      KLLA0C01254g   
8                up_up               flat      KLLA0C01298g   
9            flat_flat               flat      KLLA0C01606g   
10           flat_flat               flat      KLLA0C01650g   
11           flat_flat               flat      KLLA0C01716g   
12           flat_flat               down      KLLA0C01782g   
13           flat_flat               flat      KLLA0C01848g   
14           down_down               down      KLLA0C01870g   
15           flat_flat               flat      KLLA0C01969g   
16           flat_flat               flat      KLLA0C02343g   
17           flat_flat               flat      KLLA0C02717g   
18           down_flat               flat      KLLA0C03454g   
19           down_flat               flat      KLLA0C03542g   
20           down_down               down      KLLA0B07623g   
21           flat_flat               flat      KLLA0A06237g   
22           down_down               down      KLLA0A06336g   
23           flat_flat               flat      KLLA0A06820g   
24             up_down               flat      KLLA0B07205g   
25           flat_flat               flat      KLLA0B07370g   
26             up_flat               flat      KLLA0B07392g   
27           flat_flat               flat      KLLA0A05775g   
28           flat_flat               down      KLLA0A05731g   
29             up_flat               flat      KLLA0A05709g   
..                 ...                ...               ...   
489          flat_flat               flat      KLLA0E24025g   
490          down_down               down      KLLA0E24047g   
491          flat_flat               flat      KLLA0E24201g   
492          flat_flat               flat      KLLA0E24531g   
493            up_flat               flat      KLLA0E24597g   
494          flat_flat               flat      KLLA0F00726g   
495            up_flat               down      KLLA0F01034g   
496          flat_flat               down      KLLA0D07073g   
497          flat_flat               flat      KLLA0D07161g   
498          flat_flat               flat      KLLA0D07216g   
499          flat_flat               flat      KLLA0D07282g   
500          flat_flat               flat      KLLA0D07304g   
501          flat_flat               flat      KLLA0D07348g   
502          down_down               down      KLLA0D07405g   
503          flat_flat               flat      KLLA0D07480g   
504          flat_flat               flat      KLLA0C10054g   
505          flat_flat               flat      KLLA0C09735g   
506          down_down               down      KLLA0B08173g   
507          flat_flat               flat      KLLA0B13024g   
508          flat_flat               flat      KLLA0F24981g   
509          flat_flat               flat      KLLA0F08723g   
510          flat_flat               flat      KLLA0B12936g   
511            up_flat                 up      KLLA0E11265g   
512          flat_flat               flat      KLLA0E11199g   
513            up_flat               flat      KLLA0D01903g   
514          flat_flat               flat      KLLA0D01683g   
515          flat_flat               flat      KLLA0D01793g   
516          flat_flat               flat      KLLA0C06248g   
517          flat_flat               flat      KLLA0C16874g   
518          flat_flat               flat      KLLA0C16918g   

     log2FoldChange_KL_gene1   Gene low Gene Name low  log2FoldChange_SC_low  \
0                